<!doctype html><html lang=en class=no-js> <head><meta charset=utf-8><meta name=viewport content="width=device-width,initial-scale=1"><meta name=description content="An awesome OpenOmics cell-seek pipeline"><meta name=author content="Skyler Kuhn"><link href=../unlock/ rel=prev><link href=../cache/ rel=next><link rel=icon href=../../assets/favicon/favicon.ico><meta name=generator content="mkdocs-1.6.1, mkdocs-material-9.7.1"><title>cell-seek shinycell - cell-seek</title><link rel=stylesheet href=../../assets/stylesheets/main.484c7ddc.min.css><link rel=stylesheet href=../../assets/stylesheets/palette.ab4e12ef.min.css><link rel=preconnect href=https://fonts.gstatic.com crossorigin><link rel=stylesheet href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback"><style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style><link rel=stylesheet href=../../css/extra.css><script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script></head> <body dir=ltr data-md-color-scheme=default data-md-color-primary=indigo data-md-color-accent=indigo> <input class=md-toggle data-md-toggle=drawer type=checkbox id=__drawer autocomplete=off> <input class=md-toggle data-md-toggle=search type=checkbox id=__search autocomplete=off> <label class=md-overlay for=__drawer></label> <div data-md-component=skip> <a href=#cell-seek-shinycell class=md-skip> Skip to content </a> </div> <div data-md-component=announce> </div> <div data-md-color-scheme=default data-md-component=outdated hidden> </div> <header class=md-header data-md-component=header> <nav class="md-header__inner md-grid" aria-label=Header> <a href=../.. title=cell-seek class="md-header__button md-logo" aria-label=cell-seek data-md-component=logo> <img src=../../assets/icons/doc-book.svg alt=logo> </a> <label class="md-header__button md-icon" for=__drawer> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg> </label> <div class=md-header__title data-md-component=header-title> <div class=md-header__ellipsis> <div class=md-header__topic> <span class=md-ellipsis> cell-seek </span> </div> <div class=md-header__topic data-md-component=header-topic> <span class=md-ellipsis> cell-seek shinycell </span> </div> </div> </div> <form class=md-header__option data-md-component=palette> <input class=md-option data-md-color-media data-md-color-scheme=default data-md-color-primary=indigo data-md-color-accent=indigo aria-label="Switch to dark mode" type=radio name=__palette id=__palette_0> <label class="md-header__button md-icon" title="Switch to dark mode" for=__palette_1 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 6a6 6 0 0 1 6 6c0 2.22-1.21 4.16-3 5.2V19a1 1 0 0 1-1 1h-4a1 1 0 0 1-1-1v-1.8c-1.79-1.04-3-2.98-3-5.2a6 6 0 0 1 6-6m2 15v1a1 1 0 0 1-1 1h-2a1 1 0 0 1-1-1v-1zm6-10h3v2h-3zM1 11h3v2H1zM13 1v3h-2V1zM4.92 3.5l2.13 2.14-1.42 1.41L3.5 4.93zm12.03 2.13 2.12-2.13 1.43 1.43-2.13 2.12z"/></svg> </label> <input class=md-option data-md-color-media data-md-color-scheme=slate data-md-color-primary=indigo data-md-color-accent=indigo aria-label="Switch to light mode" type=radio name=__palette id=__palette_1> <label class="md-header__button md-icon" title="Switch to light mode" for=__palette_0 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="m17.75 4.09-2.53 1.94.91 3.06-2.63-1.81-2.63 1.81.91-3.06-2.53-1.94L12.44 4l1.06-3 1.06 3zm3.5 6.91-1.64 1.25.59 1.98-1.7-1.17-1.7 1.17.59-1.98L15.75 11l2.06-.05L18.5 9l.69 1.95zm-2.28 4.95c.83-.08 1.72 1.1 1.19 1.85-.32.45-.66.87-1.08 1.27C15.17 23 8.84 23 4.94 19.07c-3.91-3.9-3.91-10.24 0-14.14.4-.4.82-.76 1.27-1.08.75-.53 1.93.36 1.85 1.19-.27 2.86.69 5.83 2.89 8.02a9.96 9.96 0 0 0 8.02 2.89m-1.64 2.02a12.08 12.08 0 0 1-7.8-3.47c-2.17-2.19-3.33-5-3.49-7.82-2.81 3.14-2.7 7.96.31 10.98 3.02 3.01 7.84 3.12 10.98.31"/></svg> </label> </form> <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script> <label class="md-header__button md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg> </label> <div class=md-search data-md-component=search role=dialog> <label class=md-search__overlay for=__search></label> <div class=md-search__inner role=search> <form class=md-search__form name=search> <input type=text class=md-search__input name=query aria-label=Search placeholder=Search autocapitalize=off autocorrect=off autocomplete=off spellcheck=false data-md-component=search-query required> <label class="md-search__icon md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg> </label> <nav class=md-search__options aria-label=Search> <button type=reset class="md-search__icon md-icon" title=Clear aria-label=Clear tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg> </button> </nav> </form> <div class=md-search__output> <div class=md-search__scrollwrap tabindex=0 data-md-scrollfix> <div class=md-search-result data-md-component=search-result> <div class=md-search-result__meta> Initializing search </div> <ol class=md-search-result__list role=presentation></ol> </div> </div> </div> </div> </div> <div class=md-header__source> <a href=https://github.com/OpenOmics/cell-seek title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg> </div> <div class=md-source__repository> OpenOmics/cell-seek </div> </a> </div> </nav> </header> <div class=md-container data-md-component=container> <nav class=md-tabs aria-label=Tabs data-md-component=tabs> <div class=md-grid> <ul class=md-tabs__list> <li class=md-tabs__item> <a href=../.. class=md-tabs__link> About </a> </li> <li class="md-tabs__item md-tabs__item--active"> <a href=../run/ class=md-tabs__link> Commands </a> </li> <li class=md-tabs__item> <a href=../../faq/questions/ class=md-tabs__link> FAQ </a> </li> <li class=md-tabs__item> <a href=../../license/ class=md-tabs__link> License </a> </li> </ul> </div> </nav> <main class=md-main data-md-component=main> <div class="md-main__inner md-grid"> <div class="md-sidebar md-sidebar--primary" data-md-component=sidebar data-md-type=navigation> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--primary md-nav--lifted md-nav--integrated" aria-label=Navigation data-md-level=0> <label class=md-nav__title for=__drawer> <a href=../.. title=cell-seek class="md-nav__button md-logo" aria-label=cell-seek data-md-component=logo> <img src=../../assets/icons/doc-book.svg alt=logo> </a> cell-seek </label> <div class=md-nav__source> <a href=https://github.com/OpenOmics/cell-seek title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg> </div> <div class=md-source__repository> OpenOmics/cell-seek </div> </a> </div> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../.. class=md-nav__link> <span class=md-ellipsis> About </span> </a> </li> <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2 checked> <label class=md-nav__link for=__nav_2 id=__nav_2_label tabindex> <span class=md-ellipsis> Commands </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_2_label aria-expanded=true> <label class=md-nav__title for=__nav_2> <span class="md-nav__icon md-icon"></span> Commands </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../run/ class=md-nav__link> <span class=md-ellipsis> cell-seek run </span> </a> </li> <li class=md-nav__item> <a href=../genome/ class=md-nav__link> <span class=md-ellipsis> cell-seek genome </span> </a> </li> <li class=md-nav__item> <a href=../unlock/ class=md-nav__link> <span class=md-ellipsis> cell-seek unlock </span> </a> </li> <li class="md-nav__item md-nav__item--active"> <input class="md-nav__toggle md-toggle" type=checkbox id=__toc> <label class="md-nav__link md-nav__link--active" for=__toc> <span class=md-ellipsis> cell-seek shinycell </span> <span class="md-nav__icon md-icon"></span> </label> <a href=./ class="md-nav__link md-nav__link--active"> <span class=md-ellipsis> cell-seek shinycell </span> </a> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#1-about class=md-nav__link> <span class=md-ellipsis> 1. About </span> </a> </li> <li class=md-nav__item> <a href=#2-synopsis class=md-nav__link> <span class=md-ellipsis> 2. Synopsis </span> </a> <nav class=md-nav aria-label="2. Synopsis"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#21-required-arguments class=md-nav__link> <span class=md-ellipsis> 2.1 Required Arguments </span> </a> </li> <li class=md-nav__item> <a href=#22-analysis-options class=md-nav__link> <span class=md-ellipsis> 2.2 Analysis Options </span> </a> </li> <li class=md-nav__item> <a href=#23-orchestration-options class=md-nav__link> <span class=md-ellipsis> 2.3 Orchestration Options </span> </a> </li> <li class=md-nav__item> <a href=#24-miscellaneous-options class=md-nav__link> <span class=md-ellipsis> 2.4 Miscellaneous Options </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#3-example class=md-nav__link> <span class=md-ellipsis> 3. Example </span> </a> <nav class=md-nav aria-label="3. Example"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#31-basic-shinycell2-application class=md-nav__link> <span class=md-ellipsis> 3.1 Basic ShinyCell2 Application </span> </a> </li> <li class=md-nav__item> <a href=#32-including-marker-genes class=md-nav__link> <span class=md-ellipsis> 3.2 Including Marker Genes </span> </a> </li> <li class=md-nav__item> <a href=#33-advanced-configuration class=md-nav__link> <span class=md-ellipsis> 3.3 Advanced Configuration </span> </a> </li> <li class=md-nav__item> <a href=#34-local-execution-mode class=md-nav__link> <span class=md-ellipsis> 3.4 Local Execution Mode </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#4-marker-file-format class=md-nav__link> <span class=md-ellipsis> 4. Marker File Format </span> </a> <nav class=md-nav aria-label="4. Marker File Format"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#41-required-columns class=md-nav__link> <span class=md-ellipsis> 4.1 Required Columns </span> </a> </li> <li class=md-nav__item> <a href=#42-example-markers-tsv-file-format-from-seurat-findmarkersallmarkers class=md-nav__link> <span class=md-ellipsis> 4.2 Example --markers TSV file format (from Seurat FindMarkers/AllMarkers) </span> </a> </li> <li class=md-nav__item> <a href=#43-findallmarkers-vs-findmarkers class=md-nav__link> <span class=md-ellipsis> 4.3 FindAllMarkers vs FindMarkers </span> </a> <nav class=md-nav aria-label="4.3 FindAllMarkers vs FindMarkers"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#findallmarkers-recommended class=md-nav__link> <span class=md-ellipsis> FindAllMarkers (Recommended) </span> </a> </li> <li class=md-nav__item> <a href=#findmarkers-requires-post-processing class=md-nav__link> <span class=md-ellipsis> FindMarkers (Requires Post-Processing) </span> </a> </li> <li class=md-nav__item> <a href=#key-differences-summary class=md-nav__link> <span class=md-ellipsis> Key Differences Summary </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#44-usage-example class=md-nav__link> <span class=md-ellipsis> 4.4 Usage Example </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#5-notes class=md-nav__link> <span class=md-ellipsis> 5. Notes </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=../cache/ class=md-nav__link> <span class=md-ellipsis> cell-seek cache </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_3> <label class=md-nav__link for=__nav_3 id=__nav_3_label tabindex=0> <span class=md-ellipsis> FAQ </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_3_label aria-expanded=false> <label class=md-nav__title for=__nav_3> <span class="md-nav__icon md-icon"></span> FAQ </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../faq/questions/ class=md-nav__link> <span class=md-ellipsis> General Questions </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=../../license/ class=md-nav__link> <span class=md-ellipsis> License </span> </a> </li> </ul> </nav> </div> </div> </div> <div class=md-content data-md-component=content> <article class="md-content__inner md-typeset"> <h1 id=cell-seek-shinycell><code>cell-seek <b>shinycell</b></code><a class=headerlink href=#cell-seek-shinycell title="Permanent link">&para;</a></h1> <h2 id=1-about>1. About<a class=headerlink href=#1-about title="Permanent link">&para;</a></h2> <p>The <code>cell-seek</code> executable is composed of several inter-related sub commands. Please see <code>cell-seek -h</code> for all available options.</p> <p>This part of the documentation describes options and concepts for <code>cell-seek <b>shinycell</b></code> sub command in more detail. With minimal configuration, the <strong><code>shinycell</code></strong> sub command enables you to build a ShinyCell2 application from a Seurat RDS object.</p> <p>The ShinyCell2 application creates an interactive web application for exploring single-cell data. The command validates inputs, creates configuration files, and writes the files and/or Shiny code required to deploy an interactive single-cell web app.</p> <p>You can find more information about ShinyCell2 at its <a href=https://github.com/the-ouyang-lab/ShinyCell2>GitHub repository</a> and the OpenOmics fork of ShinyCell2 <a href=https://github.com/OpenOmics/ShinyCell2>here</a>. Setting up the ShinyCell2 pipeline is fast and easy! In its most basic form, <code>cell-seek <b>shinycell</b></code> only has <em>3 required inputs</em>:</p> <ul> <li>Seurat RDS object file. (with a set of valid cluster identities labels) </li> <li>A project title.</li> <li>An output directory to write the shiny application files.</li> </ul> <p>A user can optionally provide a TSV file with marker genes to add a DEG page to the application.</p> <div class="admonition important"> <p class=admonition-title>Supported Assays</p> <p><code>ShinyCell2</code> does not support all <code>Seurat</code> object assays. <strong>Mileage may vary with the Seurat Object you use.</strong> <strong><em>HTO assays</em></strong> and <strong><em>Azimuth cell.annotation</em></strong> <code>(prediction.score.celltype.l*)</code> assays will be removed from your object when using the ShinyCell2 pipeline.</p> <p>These assay types are unsupported currently by the original ShinyCell2.</p> <p>For more information about support for assays and extended format support please consult the original ShinyCell2 repository: <a href=https://github.com/the-ouyang-lab/ShinyCell2>https://github.com/the-ouyang-lab/ShinyCell2</a></p> </div> <h2 id=2-synopsis>2. Synopsis<a class=headerlink href=#2-synopsis title="Permanent link">&para;</a></h2> <div class=highlight><pre><span></span><code>$ cell-seek shinycell [-h] \
      -j OBJ -o OUTDIR --proj PROJECT \
      [--markers MARKERS] [--cluster_labels CLUSTER_LABELS] \
      [--rmmeta META1,META2] [--defred DEFAULTRED] \
      [-l MAXLEVELS] [-a ASSAYS] \
      [--mode {slurm,local}] [--threads THREADS] \
      [--tmp-dir TMP_DIR] [--job-name JOB_NAME] [--silent]
</code></pre></div> <p>The synopsis for this command shows its arguments and their usage. Optional arguments are shown in square brackets.</p> <p>A user <strong>must</strong> provide a path to a Seurat RDS object via <code>-j</code> or <code>--obj</code> argument, an output directory to store results via <code>-o</code> or <code>--outdir</code> argument, and a project name via <code>--proj</code> argument.</p> <p>You can always use the <code>-h</code> option for information on a specific command.</p> <h3 id=21-required-arguments>2.1 Required Arguments<a class=headerlink href=#21-required-arguments title="Permanent link">&para;</a></h3> <p>Each of the following arguments are required. Failure to provide a required argument will result in a non-zero exit-code.</p> <p><code>-j OBJ, --obj OBJ</code> </p> <blockquote> <p><strong>Path to Seurat RDS object file.</strong><br> <em>type: file</em> </p> <p>Path to an RDS file produced with saveRDS() that contains a Seurat object. The script will attempt to read and validate the Seurat object.</p> <p><strong><em>Example:</em></strong> <code>-j seurat_object.rds</code></p> </blockquote> <hr> <p><code>-o OUTDIR, --outdir OUTDIR</code></p> <blockquote> <p><strong>Path to an output directory.</strong> <br> <em>type: path</em></p> <p>Output directory where the Shiny app files and generated code will be written. Directory will be created if it does not exist.</p> <p><strong><em>Example:</em></strong> <code>-o /srv/shiny-server/my_app</code></p> </blockquote> <hr> <p><code>--proj PROJECT</code></p> <blockquote> <p><strong>Project name for the application.</strong> <br> <em>type: string</em></p> <p>Project name used as the app title and prefix for generated assets.</p> <p><strong><em>Example:</em></strong> <code>--proj NCBR-34</code></p> </blockquote> <h3 id=22-analysis-options>2.2 Analysis Options<a class=headerlink href=#22-analysis-options title="Permanent link">&para;</a></h3> <p>Each of the following arguments are optional, and do not need to be provided.</p> <p><code>--markers MARKERS</code></p> <blockquote> <p><strong>Path to a marker genes file.</strong> <br> <em>type: file</em></p> <p>Path to a plain text file with marker genes, one gene per line. When provided, requires <code>--cluster_labels</code> to be set as well. Used to generate marker-based DEG pages or panels. If not provided, no DEG pages/panels will be created.</p> <p><strong><em>Example:</em></strong> <code>--markers markers.txt</code></p> </blockquote> <hr> <p><code>--cluster_labels CLUSTER_LABELS</code></p> <blockquote> <p><strong>Metadata column name for cluster identities.</strong> <br> <em>type: string</em></p> <p>Metadata column name in <a href=mailto:seurat_obj@meta.data>&#115;&#101;&#117;&#114;&#97;&#116;&#95;&#111;&#98;&#106;&#64;&#109;&#101;&#116;&#97;&#46;&#100;&#97;&#116;&#97;</a> that contains cluster identities corresponding to the markers file. Required when <code>--markers</code> is provided. If not provided along with markers, no DEG pages/panels will be created.</p> <p><strong><em>Example:</em></strong> <code>--cluster_labels seurat_clusters</code></p> </blockquote> <hr> <p><code>--rmmeta META1,META2</code></p> <blockquote> <p><strong>Metadata columns to remove.</strong> <br> <em>type: string</em></p> <p>Comma-delimited list of metadata column names to drop from the Seurat object before deploying. Useful to remove large or sensitive columns. If not provided, no metadata columns are removed.</p> <p><strong><em>Example:</em></strong> <code>--rmmeta barcode,unwanted_col</code></p> </blockquote> <hr> <p><code>--defred DEFAULTRED</code></p> <blockquote> <p><strong>Default dimensionality reduction.</strong> <br> <em>type: string</em></p> <p>Default dimensionality reduction to use for the app (must exist in Seurat object reductions). The code expects the reduction to expose a key so UMAP1/UMAP2 (or <key>1/<key>2) coordinate names can be derived. If not provided, uses all active reductions.</p> <p><strong><em>Example:</em></strong> <code>--defred UMAP</code></p> </blockquote> <hr> <p><code>-l MAXLEVELS, --maxlevels MAXLEVELS</code></p> <blockquote> <p><strong>Maximum levels for categorical metadata.</strong> <br> <em>type: int</em><br> <em>default: 50</em></p> <p>Maximum allowed number of factor levels for categorical metadata fields. Fields exceeding this will be treated as continuous / filtered out.</p> <p><strong><em>Example:</em></strong> <code>-l 50</code></p> </blockquote> <hr> <p><code>-a ASSAYS, --assay ASSAYS</code></p> <blockquote> <p><strong>Assays to include in the app.</strong> <br> <em>type: string</em></p> <p>Comma-delimited list of assays to include in the Shiny app (e.g. RNA,spatial,ATAC). If omitted, default assays present in the Seurat object are used when compatible.</p> <p><strong><em>Example:</em></strong> <code>-a RNA,spatial</code></p> </blockquote> <h3 id=23-orchestration-options>2.3 Orchestration Options<a class=headerlink href=#23-orchestration-options title="Permanent link">&para;</a></h3> <p>Each of the following arguments are optional, and do not need to be provided.</p> <p><code>--mode {slurm,local}</code> </p> <blockquote> <p><strong>Execution Method.</strong><br> <em>type: string</em><br> <em>default: slurm</em></p> <p>Execution Method. Defines the mode or method of execution. Valid mode options include: slurm or local.</p> <p><strong><em>slurm</em></strong> <br> The slurm execution method will submit jobs to the <a href=https://slurm.schedmd.com/ >SLURM workload manager</a>. It is recommended running cell-seek in this mode as execution will be significantly faster in a distributed environment. This is the default mode of execution.</p> <p><strong><em>local</em></strong><br> Local executions will run serially on compute instance. This is useful for testing, debugging, or when a users does not have access to a high performance computing environment. If this option is not provided, it will default to a local execution mode.</p> <p><strong><em>Example:</em></strong> <code>--mode slurm</code></p> </blockquote> <hr> <p><code>--job-name JOB_NAME</code> </p> <blockquote> <p><strong>Set the name of the pipeline's master job.</strong><br> <em>type: string</em> <em>default: pl:cell-seek:shinycell</em></p> <p>When submitting the pipeline to a job scheduler, like SLURM, this option allows you to set the name of the pipeline's master job. By default, the name of the pipeline's master job is set to "pl:cell-seek:shinycell".</p> <p><strong><em>Example:</em></strong> <code>--job-name pl_shinycell_id-42</code></p> </blockquote> <hr> <p><code>--silent</code> </p> <blockquote> <p><strong>Silence standard output.</strong><br> <em>type: boolean flag</em></p> <p>Reduces the amount of information directed to standard output when submitting master job to the job scheduler. Only the job id of the master job is returned.</p> <p><strong><em>Example:</em></strong> <code>--silent</code></p> </blockquote> <hr> <p><code>--threads THREADS</code> </p> <blockquote> <p><strong>Max number of threads for each process.</strong><br> <em>type: int</em><br> <em>default: 2</em></p> <p>Max number of threads for each process. This option is more applicable when running the pipeline with <code>--mode local</code>. It is recommended setting this value to the maximum number of CPUs available on the host machine.</p> <p><strong><em>Example:</em></strong> <code>--threads 12</code></p> </blockquote> <hr> <p><code>--tmp-dir TMP_DIR</code> </p> <blockquote> <p><strong>Path for temporary files.</strong><br> <em>type: path</em><br> <em>default: <code>/lscratch/$SLURM_JOBID</code></em></p> <p>Path on the file system for writing temporary output files. By default, the temporary directory is set to '/lscratch/$SLURM_JOBID' for backwards compatibility with the NIH's Biowulf cluster; however, if you are running the pipeline on another cluster, this option will need to be specified. Ideally, this path should point to a dedicated location on the filesystem for writing tmp files. On many systems, this location is set to somewhere in /scratch. If you need to inject a variable into this string that should NOT be expanded, please quote this options value in single quotes.</p> <p><strong><em>Example:</em></strong> <code>--tmp-dir /scratch/$USER/</code></p> </blockquote> <h3 id=24-miscellaneous-options>2.4 Miscellaneous Options<a class=headerlink href=#24-miscellaneous-options title="Permanent link">&para;</a></h3> <p>Each of the following arguments are optional, and do not need to be provided.</p> <p><code>-h, --help</code> </p> <blockquote> <p><strong>Display Help.</strong><br> <em>type: boolean flag</em></p> <p>Shows command's synopsis, help message, and an example command</p> <p><strong><em>Example:</em></strong> <code>--help</code></p> </blockquote> <h2 id=3-example>3. Example<a class=headerlink href=#3-example title="Permanent link">&para;</a></h2> <h3 id=31-basic-shinycell2-application>3.1 Basic ShinyCell2 Application<a class=headerlink href=#31-basic-shinycell2-application title="Permanent link">&para;</a></h3> <p>Create a basic ShinyCell2 app from a Seurat RDS file:</p> <div class=highlight><pre><span></span><code><span class=c1># Step 1.) Grab an interactive node,</span>
<span class=c1># do not run on head node!</span>
srun<span class=w> </span>-N<span class=w> </span><span class=m>1</span><span class=w> </span>-n<span class=w> </span><span class=m>1</span><span class=w> </span>--time<span class=o>=</span><span class=m>1</span>:00:00<span class=w> </span>--mem<span class=o>=</span>8gb<span class=w> </span>--cpus-per-task<span class=o>=</span><span class=m>2</span><span class=w> </span>--pty<span class=w> </span>bash
module<span class=w> </span>purge
module<span class=w> </span>load<span class=w> </span>singularity<span class=w> </span>snakemake

<span class=c1># Step 2.) Run the shinycell subcommand</span>
./cell-seek<span class=w> </span>shinycell<span class=w> </span>-j<span class=w> </span>seurat_obj.rds<span class=w> </span><span class=se>\</span>
<span class=w>                       </span>-o<span class=w> </span>./ShinyCell2_app<span class=w> </span><span class=se>\</span>
<span class=w>                       </span>--proj<span class=w> </span>NCBR-34<span class=w> </span><span class=se>\</span>
<span class=w>                       </span>--mode<span class=w> </span>slurm
</code></pre></div> <h3 id=32-including-marker-genes>3.2 Including Marker Genes<a class=headerlink href=#32-including-marker-genes title="Permanent link">&para;</a></h3> <p>Create a ShinyCell2 app with marker genes and cluster labels:</p> <div class=highlight><pre><span></span><code><span class=c1># Step 1.) Grab an interactive node,</span>
<span class=c1># do not run on head node!</span>
srun<span class=w> </span>-N<span class=w> </span><span class=m>1</span><span class=w> </span>-n<span class=w> </span><span class=m>1</span><span class=w> </span>--time<span class=o>=</span><span class=m>1</span>:00:00<span class=w> </span>--mem<span class=o>=</span>8gb<span class=w> </span>--cpus-per-task<span class=o>=</span><span class=m>2</span><span class=w> </span>--pty<span class=w> </span>bash
module<span class=w> </span>purge
module<span class=w> </span>load<span class=w> </span>singularity<span class=w> </span>snakemake

<span class=c1># Step 2.) Run the shinycell subcommand with markers</span>
./cell-seek<span class=w> </span>shinycell<span class=w> </span>-j<span class=w> </span>seurat_obj.rds<span class=w> </span><span class=se>\</span>
<span class=w>                       </span>-o<span class=w> </span>./ShinyCell2_markers<span class=w> </span><span class=se>\</span>
<span class=w>                       </span>--proj<span class=w> </span>NCBR-34<span class=w> </span><span class=se>\</span>
<span class=w>                       </span>--markers<span class=w> </span>markers.txt<span class=w> </span><span class=se>\</span>
<span class=w>                       </span>--cluster_labels<span class=w> </span>seurat_clusters<span class=w> </span><span class=se>\</span>
<span class=w>                       </span>--mode<span class=w> </span>slurm
</code></pre></div> <h3 id=33-advanced-configuration>3.3 Advanced Configuration<a class=headerlink href=#33-advanced-configuration title="Permanent link">&para;</a></h3> <p>Create a ShinyCell2 app with specific assays and metadata filtering:</p> <div class=highlight><pre><span></span><code><span class=c1># Step 1.) Grab an interactive node,</span>
<span class=c1># do not run on head node!</span>
srun<span class=w> </span>-N<span class=w> </span><span class=m>1</span><span class=w> </span>-n<span class=w> </span><span class=m>1</span><span class=w> </span>--time<span class=o>=</span><span class=m>1</span>:00:00<span class=w> </span>--mem<span class=o>=</span>8gb<span class=w> </span>--cpus-per-task<span class=o>=</span><span class=m>2</span><span class=w> </span>--pty<span class=w> </span>bash
module<span class=w> </span>purge
module<span class=w> </span>load<span class=w> </span>singularity<span class=w> </span>snakemake

<span class=c1># Step 2.) Run the shinycell subcommand with advanced options</span>
./cell-seek<span class=w> </span>shinycell<span class=w> </span>-j<span class=w> </span>seurat_obj.rds<span class=w> </span><span class=se>\</span>
<span class=w>                       </span>-o<span class=w> </span>./ShinyCell2_advanced<span class=w> </span><span class=se>\</span>
<span class=w>                       </span>--proj<span class=w> </span>NCBR-34<span class=w> </span><span class=se>\</span>
<span class=w>                       </span>-a<span class=w> </span>RNA,spatial<span class=w> </span><span class=se>\</span>
<span class=w>                       </span>--rmmeta<span class=w> </span>large_matrix,sensitive_col<span class=w> </span><span class=se>\</span>
<span class=w>                       </span>--defred<span class=w> </span>UMAP<span class=w> </span><span class=se>\</span>
<span class=w>                       </span>-l<span class=w> </span><span class=m>100</span><span class=w> </span><span class=se>\</span>
<span class=w>                       </span>--mode<span class=w> </span>slurm
</code></pre></div> <h3 id=34-local-execution-mode>3.4 Local Execution Mode<a class=headerlink href=#34-local-execution-mode title="Permanent link">&para;</a></h3> <p>Run ShinyCell2 in local mode for testing or debugging:</p> <div class=highlight><pre><span></span><code><span class=c1># Step 1.) Ensure you have sufficient resources</span>
<span class=c1># on your current node</span>

<span class=c1># Step 2.) Run the shinycell subcommand in local mode</span>
./cell-seek<span class=w> </span>shinycell<span class=w> </span>-j<span class=w> </span>seurat_obj.rds<span class=w> </span><span class=se>\</span>
<span class=w>                       </span>-o<span class=w> </span>./ShinyCell2_local<span class=w> </span><span class=se>\</span>
<span class=w>                       </span>--proj<span class=w> </span>NCBR-34<span class=w> </span><span class=se>\</span>
<span class=w>                       </span>--mode<span class=w> </span><span class=nb>local</span><span class=w> </span><span class=se>\</span>
<span class=w>                       </span>--threads<span class=w> </span><span class=m>8</span>
</code></pre></div> <h2 id=4-marker-file-format>4. Marker File Format<a class=headerlink href=#4-marker-file-format title="Permanent link">&para;</a></h2> <p>The <code>--markers</code> option expects a TSV (tab-separated values) file in the same format as produced by Seurat's <code>FindAllMarkers()</code> function. This file contains differential expression results for marker genes across clusters.</p> <h3 id=41-required-columns>4.1 Required Columns<a class=headerlink href=#41-required-columns title="Permanent link">&para;</a></h3> <p>The marker file must contain the following columns:</p> <ul> <li><strong>cluster</strong>: The cluster identity (must match values in the metadata column specified by <code>--cluster_labels</code>)</li> <li><strong>gene</strong>: Gene name or identifier</li> <li><strong>p_val</strong>: Unadjusted p-value</li> <li><strong>avg_log2FC</strong>: Average log2 fold-change</li> <li><strong>pct.1</strong>: Percentage of cells expressing the gene in the cluster</li> <li><strong>pct.2</strong>: Percentage of cells expressing the gene in all other clusters</li> <li><strong>p_val_adj</strong>: Adjusted p-value (e.g., Bonferroni correction)</li> </ul> <h3 id=42-example-markers-tsv-file-format-from-seurat-findmarkersallmarkers>4.2 Example <code>--markers</code> TSV file format (from Seurat FindMarkers/AllMarkers)<a class=headerlink href=#42-example-markers-tsv-file-format-from-seurat-findmarkersallmarkers title="Permanent link">&para;</a></h3> <table> <thead> <tr> <th>p_val</th> <th>avg_log2FC</th> <th>pct.1</th> <th>pct.2</th> <th>p_val_adj</th> <th>cluster</th> <th>gene</th> </tr> </thead> <tbody> <tr> <td>0</td> <td>2.5</td> <td>0.95</td> <td>0.25</td> <td>0</td> <td>0</td> <td>CD3D</td> </tr> <tr> <td>0</td> <td>2.3</td> <td>0.92</td> <td>0.18</td> <td>0</td> <td>0</td> <td>CD3E</td> </tr> <tr> <td>1.2E-250</td> <td>1.8</td> <td>0.85</td> <td>0.35</td> <td>2.4E-246</td> <td>1</td> <td>CD14</td> </tr> </tbody> </table> <h3 id=43-findallmarkers-vs-findmarkers>4.3 FindAllMarkers vs FindMarkers<a class=headerlink href=#43-findallmarkers-vs-findmarkers title="Permanent link">&para;</a></h3> <p>Seurat provides two main functions for identifying marker genes:</p> <h4 id=findallmarkers-recommended>FindAllMarkers (Recommended)<a class=headerlink href=#findallmarkers-recommended title="Permanent link">&para;</a></h4> <p><code>FindAllMarkers()</code> automatically identifies markers for <strong>all clusters</strong> and includes the <code>cluster</code> column in its output, making it directly compatible with ShinyCell2:</p> <div class=highlight><pre><span></span><code><span class=c1># Run FindAllMarkers - produces ShinyCell2-compatible output</span>
<span class=n>markers</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>FindAllMarkers</span><span class=p>(</span><span class=n>seurat_obj</span><span class=p>,</span><span class=w> </span>
<span class=w>                         </span><span class=n>only.pos</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=kc>TRUE</span><span class=p>,</span>
<span class=w>                         </span><span class=n>min.pct</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>0.25</span><span class=p>,</span>
<span class=w>                         </span><span class=n>logfc.threshold</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>0.25</span><span class=p>)</span>

<span class=c1># Save directly - cluster column is already included</span>
<span class=nf>write.table</span><span class=p>(</span><span class=n>markers</span><span class=p>,</span><span class=w> </span>
<span class=w>           </span><span class=n>file</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;markers.tsv&quot;</span><span class=p>,</span>
<span class=w>           </span><span class=n>sep</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;\t&quot;</span><span class=p>,</span>
<span class=w>           </span><span class=n>quote</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=kc>FALSE</span><span class=p>,</span>
<span class=w>           </span><span class=n>row.names</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=kc>FALSE</span><span class=p>)</span>
</code></pre></div> <h4 id=findmarkers-requires-post-processing>FindMarkers (Requires Post-Processing)<a class=headerlink href=#findmarkers-requires-post-processing title="Permanent link">&para;</a></h4> <p><code>FindMarkers()</code> compares gene expression between <strong>two specific groups</strong> but does <strong>not</strong> include cluster labels in its output. To use <code>FindMarkers()</code> results with ShinyCell2, you must manually add the <code>cluster</code> column.</p> <p>The following script demonstrates how to run <code>FindMarkers()</code> on each cluster and create <code>FindAllMarkers</code>-like output:</p> <div class=highlight><pre><span></span><code><span class=c1>#!/usr/bin/env Rscript</span>
<span class=c1># generate_markers_from_findmarkers.R</span>
<span class=c1># Creates FindAllMarkers-compatible output using FindMarkers for each cluster</span>

<span class=nf>library</span><span class=p>(</span><span class=n>Seurat</span><span class=p>)</span>

<span class=c1># Load your Seurat object</span>
<span class=n>seurat_obj</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>readRDS</span><span class=p>(</span><span class=s>&quot;seurat_object.rds&quot;</span><span class=p>)</span>

<span class=c1># Specify the metadata column containing cluster identities</span>
<span class=n>cluster_column</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=s>&quot;seurat_clusters&quot;</span><span class=w>  </span><span class=c1># Change this to match your metadata column</span>

<span class=c1># Get unique cluster identities</span>
<span class=n>clusters</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>unique</span><span class=p>(</span><span class=n>seurat_obj</span><span class=o>@</span><span class=n>meta.data</span><span class=p>[[</span><span class=n>cluster_column</span><span class=p>]])</span>

<span class=c1># Initialize an empty list to store results</span>
<span class=n>all_markers</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>list</span><span class=p>()</span>

<span class=c1># Loop through each cluster and run FindMarkers</span>
<span class=kr>for</span><span class=w> </span><span class=p>(</span><span class=n>cluster_id</span><span class=w> </span><span class=kr>in</span><span class=w> </span><span class=n>clusters</span><span class=p>)</span><span class=w> </span><span class=p>{</span>
<span class=w>  </span><span class=nf>cat</span><span class=p>(</span><span class=nf>paste0</span><span class=p>(</span><span class=s>&quot;Processing cluster: &quot;</span><span class=p>,</span><span class=w> </span><span class=n>cluster_id</span><span class=p>,</span><span class=w> </span><span class=s>&quot;\n&quot;</span><span class=p>))</span>

<span class=w>  </span><span class=c1># Run FindMarkers comparing this cluster vs all others</span>
<span class=w>  </span><span class=nf>tryCatch</span><span class=p>({</span>
<span class=w>    </span><span class=n>markers</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>FindMarkers</span><span class=p>(</span>
<span class=w>      </span><span class=n>seurat_obj</span><span class=p>,</span>
<span class=w>      </span><span class=n>ident.1</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>cluster_id</span><span class=p>,</span>
<span class=w>      </span><span class=n>ident.2</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=kc>NULL</span><span class=p>,</span><span class=w>  </span><span class=c1># Compare against all other cells</span>
<span class=w>      </span><span class=n>group.by</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>cluster_column</span><span class=p>,</span>
<span class=w>      </span><span class=n>only.pos</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=kc>TRUE</span><span class=p>,</span>
<span class=w>      </span><span class=n>min.pct</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>0.25</span><span class=p>,</span>
<span class=w>      </span><span class=n>logfc.threshold</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>0.25</span><span class=p>,</span>
<span class=w>      </span><span class=n>test.use</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;wilcox&quot;</span><span class=w>  </span><span class=c1># Wilcoxon rank sum test (default)</span>
<span class=w>    </span><span class=p>)</span>

<span class=w>    </span><span class=c1># Add gene names as a column (from rownames)</span>
<span class=w>    </span><span class=n>markers</span><span class=o>$</span><span class=n>gene</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>rownames</span><span class=p>(</span><span class=n>markers</span><span class=p>)</span>

<span class=w>    </span><span class=c1># Add cluster identity column</span>
<span class=w>    </span><span class=n>markers</span><span class=o>$</span><span class=n>cluster</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=n>cluster_id</span>

<span class=w>    </span><span class=c1># Store in list</span>
<span class=w>    </span><span class=n>all_markers</span><span class=p>[[</span><span class=nf>as.character</span><span class=p>(</span><span class=n>cluster_id</span><span class=p>)]]</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=n>markers</span>

<span class=w>  </span><span class=p>},</span><span class=w> </span><span class=n>error</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=kr>function</span><span class=p>(</span><span class=n>e</span><span class=p>)</span><span class=w> </span><span class=p>{</span>
<span class=w>    </span><span class=nf>cat</span><span class=p>(</span><span class=nf>paste0</span><span class=p>(</span><span class=s>&quot;Error processing cluster &quot;</span><span class=p>,</span><span class=w> </span><span class=n>cluster_id</span><span class=p>,</span><span class=w> </span><span class=s>&quot;: &quot;</span><span class=p>,</span><span class=w> </span><span class=n>e</span><span class=o>$</span><span class=n>message</span><span class=p>,</span><span class=w> </span><span class=s>&quot;\n&quot;</span><span class=p>))</span>
<span class=w>  </span><span class=p>})</span>
<span class=p>}</span>

<span class=c1># Combine all results into a single data frame</span>
<span class=n>combined_markers</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>do.call</span><span class=p>(</span><span class=n>rbind</span><span class=p>,</span><span class=w> </span><span class=n>all_markers</span><span class=p>)</span>

<span class=c1># Reset row names</span>
<span class=nf>rownames</span><span class=p>(</span><span class=n>combined_markers</span><span class=p>)</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=kc>NULL</span>

<span class=c1># Reorder columns to match FindAllMarkers output</span>
<span class=n>combined_markers</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=n>combined_markers</span><span class=p>[,</span><span class=w> </span><span class=nf>c</span><span class=p>(</span><span class=s>&quot;p_val&quot;</span><span class=p>,</span><span class=w> </span><span class=s>&quot;avg_log2FC&quot;</span><span class=p>,</span><span class=w> </span><span class=s>&quot;pct.1&quot;</span><span class=p>,</span><span class=w> </span><span class=s>&quot;pct.2&quot;</span><span class=p>,</span><span class=w> </span>
<span class=w>                                         </span><span class=s>&quot;p_val_adj&quot;</span><span class=p>,</span><span class=w> </span><span class=s>&quot;cluster&quot;</span><span class=p>,</span><span class=w> </span><span class=s>&quot;gene&quot;</span><span class=p>)]</span>

<span class=c1># Optional: Filter by significance and fold-change</span>
<span class=n>combined_markers</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=n>combined_markers</span><span class=p>[</span><span class=n>combined_markers</span><span class=o>$</span><span class=n>p_val_adj</span><span class=w> </span><span class=o>&lt;</span><span class=w> </span><span class=m>0.05</span><span class=w> </span><span class=o>&amp;</span><span class=w> </span>
<span class=w>                                     </span><span class=n>combined_markers</span><span class=o>$</span><span class=n>avg_log2FC</span><span class=w> </span><span class=o>&gt;</span><span class=w> </span><span class=m>0.25</span><span class=p>,</span><span class=w> </span><span class=p>]</span>

<span class=c1># Optional: Sort by cluster and adjusted p-value</span>
<span class=n>combined_markers</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=n>combined_markers</span><span class=p>[</span><span class=nf>order</span><span class=p>(</span><span class=n>combined_markers</span><span class=o>$</span><span class=n>cluster</span><span class=p>,</span><span class=w> </span>
<span class=w>                                           </span><span class=n>combined_markers</span><span class=o>$</span><span class=n>p_val_adj</span><span class=p>),</span><span class=w> </span><span class=p>]</span>

<span class=c1># Save to TSV file</span>
<span class=nf>write.table</span><span class=p>(</span><span class=n>combined_markers</span><span class=p>,</span>
<span class=w>           </span><span class=n>file</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;markers_from_findmarkers.tsv&quot;</span><span class=p>,</span>
<span class=w>           </span><span class=n>sep</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;\t&quot;</span><span class=p>,</span>
<span class=w>           </span><span class=n>quote</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=kc>FALSE</span><span class=p>,</span>
<span class=w>           </span><span class=n>row.names</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=kc>FALSE</span><span class=p>)</span>

<span class=nf>cat</span><span class=p>(</span><span class=nf>paste0</span><span class=p>(</span><span class=s>&quot;\nGenerated &quot;</span><span class=p>,</span><span class=w> </span><span class=nf>nrow</span><span class=p>(</span><span class=n>combined_markers</span><span class=p>),</span><span class=w> </span><span class=s>&quot; marker entries across &quot;</span><span class=p>,</span><span class=w> </span>
<span class=w>          </span><span class=nf>length</span><span class=p>(</span><span class=nf>unique</span><span class=p>(</span><span class=n>combined_markers</span><span class=o>$</span><span class=n>cluster</span><span class=p>)),</span><span class=w> </span><span class=s>&quot; clusters\n&quot;</span><span class=p>))</span>
<span class=nf>cat</span><span class=p>(</span><span class=s>&quot;Output saved to: markers_from_findmarkers.tsv\n&quot;</span><span class=p>)</span>
</code></pre></div> <h4 id=key-differences-summary>Key Differences Summary<a class=headerlink href=#key-differences-summary title="Permanent link">&para;</a></h4> <table> <thead> <tr> <th>Feature</th> <th>FindAllMarkers</th> <th>FindMarkers</th> </tr> </thead> <tbody> <tr> <td><strong>Cluster column</strong></td> <td> Included automatically</td> <td> Must be added manually</td> </tr> <tr> <td><strong>Scope</strong></td> <td>All clusters at once</td> <td>One comparison at a time</td> </tr> <tr> <td><strong>Use case</strong></td> <td>Quick marker discovery</td> <td>Specific pairwise comparisons</td> </tr> <tr> <td><strong>ShinyCell2 compatibility</strong></td> <td> Direct</td> <td> Requires post-processing</td> </tr> </tbody> </table> <h3 id=44-usage-example>4.4 Usage Example<a class=headerlink href=#44-usage-example title="Permanent link">&para;</a></h3> <p>Once you have generated your marker file using either method:</p> <div class=highlight><pre><span></span><code>./cell-seek<span class=w> </span>shinycell<span class=w> </span>-j<span class=w> </span>seurat_obj.rds<span class=w> </span><span class=se>\</span>
<span class=w>                       </span>-o<span class=w> </span>./ShinyCell2_app<span class=w> </span><span class=se>\</span>
<span class=w>                       </span>--proj<span class=w> </span>MyProject<span class=w> </span><span class=se>\</span>
<span class=w>                       </span>--markers<span class=w> </span>markers.tsv<span class=w> </span><span class=se>\</span>
<span class=w>                       </span>--cluster_labels<span class=w> </span>seurat_clusters<span class=w> </span><span class=se>\</span>
<span class=w>                       </span>--mode<span class=w> </span>slurm
</code></pre></div> <h2 id=5-notes>5. Notes<a class=headerlink href=#5-notes title="Permanent link">&para;</a></h2> <ul> <li>When using <code>--markers</code> you <strong>MUST</strong> also specify <code>--cluster_labels</code>.</li> <li>The <code>--defred</code> option must correspond to an existing reduction name in the Seurat object; the script will resolve the reduction key to form e.g. UMAP1/UMAP2.</li> <li>Use <code>--rmmeta</code> to remove columns that are large, private, or not appropriate for public deployment.</li> </ul> <aside class=md-source-file> <span class=md-source-file__fact> <span class=md-icon title="Last update"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M21 13.1c-.1 0-.3.1-.4.2l-1 1 2.1 2.1 1-1c.2-.2.2-.6 0-.8l-1.3-1.3c-.1-.1-.2-.2-.4-.2m-1.9 1.8-6.1 6V23h2.1l6.1-6.1zM12.5 7v5.2l4 2.4-1 1L11 13V7zM11 21.9c-5.1-.5-9-4.8-9-9.9C2 6.5 6.5 2 12 2c5.3 0 9.6 4.1 10 9.3-.3-.1-.6-.2-1-.2s-.7.1-1 .2C19.6 7.2 16.2 4 12 4c-4.4 0-8 3.6-8 8 0 4.1 3.1 7.5 7.1 7.9l-.1.2z"/></svg> </span> 2026-01-08 </span> </aside> </article> </div> <script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script> </div> <button type=button class="md-top md-icon" data-md-component=top hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"/></svg> Back to top </button> </main> <footer class=md-footer> <div class="md-footer-meta md-typeset"> <div class="md-footer-meta__inner md-grid"> <div class=md-copyright> <div class=md-copyright__highlight> Copyright &copy; 2026 OpenOmics </div> Made with <a href=https://squidfunk.github.io/mkdocs-material/ target=_blank rel=noopener> Material for MkDocs </a> </div> <div class=md-social> <a href=https://ncbr.ncifcrf.gov/ target=_blank rel=noopener title=ncbr.ncifcrf.gov class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 640 512"><!-- Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M320 16a104 104 0 1 1 0 208 104 104 0 1 1 0-208M96 88a72 72 0 1 1 0 144 72 72 0 1 1 0-144M0 416c0-70.7 57.3-128 128-128 12.8 0 25.2 1.9 36.9 5.4C132 330.2 112 378.8 112 432v16c0 11.4 2.4 22.2 6.7 32H32c-17.7 0-32-14.3-32-32zm521.3 64c4.3-9.8 6.7-20.6 6.7-32v-16c0-53.2-20-101.8-52.9-138.6 11.7-3.5 24.1-5.4 36.9-5.4 70.7 0 128 57.3 128 128v32c0 17.7-14.3 32-32 32zM472 160a72 72 0 1 1 144 0 72 72 0 1 1-144 0M160 432c0-88.4 71.6-160 160-160s160 71.6 160 160v16c0 17.7-14.3 32-32 32H192c-17.7 0-32-14.3-32-32z"/></svg> </a> <a href=https://github.com/OpenOmics target=_blank rel=noopener title=github.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 512 512"><!-- Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M173.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M252.8 8C114.1 8 8 113.3 8 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C436.2 457.8 504 362.9 504 252 504 113.3 391.5 8 252.8 8M105.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg> </a> <a href=https://hub.docker.com/u/skchronicles target=_blank rel=noopener title=hub.docker.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 640 512"><!-- Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M349.9 236.3h-66.1v-59.4h66.1zm0-204.3h-66.1v60.7h66.1zm78.2 144.8H362v59.4h66.1zm-156.3-72.1h-66.1v60.1h66.1zm78.1 0h-66.1v60.1h66.1zm276.8 100c-14.4-9.7-47.6-13.2-73.1-8.4-3.3-24-16.7-44.9-41.1-63.7l-14-9.3-9.3 14c-18.4 27.8-23.4 73.6-3.7 103.8-8.7 4.7-25.8 11.1-48.4 10.7H2.4c-8.7 50.8 5.8 116.8 44 162.1 37.1 43.9 92.7 66.2 165.4 66.2 157.4 0 273.9-72.5 328.4-204.2 21.4.4 67.6.1 91.3-45.2 1.5-2.5 6.6-13.2 8.5-17.1zm-511.1-27.9h-66v59.4h66.1v-59.4zm78.1 0h-66.1v59.4h66.1zm78.1 0h-66.1v59.4h66.1zm-78.1-72.1h-66.1v60.1h66.1z"/></svg> </a> </div> </div> </div> </footer> </div> <div class=md-dialog data-md-component=dialog> <div class="md-dialog__inner md-typeset"></div> </div> <script id=__config type=application/json>{"annotate": null, "base": "../..", "features": ["navigation.tabs", "navigation.top", "toc.integrate"], "search": "../../assets/javascripts/workers/search.2c215733.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": {"provider": "mike"}}</script> <script src=../../assets/javascripts/bundle.79ae519e.min.js></script> </body> </html>