---
title: "Preliminary QC Report for Sample `r params$sample`"
subtitle: "Initial Sample Preprocessing"
date: '`r Sys.Date()`'
output:
    rmdformats::robobook:
        lightbox: true
        number_sections: true
        gallery: true
        code-fold: true
        toc_depth: 3
params:
  seuratdir: seurat
  sample: sample
  defaultfilter: True
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, message=FALSE, warning = FALSE, results='asis')
options(knitr.graphics.error = FALSE)
```

```{r Packages, message=FALSE}
library(knitr)
library(stringr)
library(flextable)
library(dplyr)
library(base64enc)
library(xfun)
# library(kableExtra)
# library(scater)
```

```{r Input, include=FALSE}
seuratdir <- params$seuratdir
```

```{r, include=FALSE}
# Init Step to make sure that the dependencies are loaded
#htmltools::tagList(DT::datatable(library_key))
#data <- read.csv('/data/NIAMS_IDSS/projects/NIAMS-33/library_patient.csv')
#htmltools::tagList(DT::datatable(data, extensions='Buttons', options=list(dom='lBfrtip', buttons = c('copy', 'csv', 'excel'), lengthMenu = list(c(10,25,50,-1), c(10,25,50,"All")))))

```

# QC Analysis

The data that was generated by Cell Ranger was pulled into Seurat for initial processing.

## Pre-Filter Violin Plots {.tabset .tabset-pills}

Violin plots of the number of genes associated with a cell, number of unique reads associated with a cell, and the percentage of counts in a cell belonging to mitochondrial genes were created. Cells with extremely low or high number of reads or counts may indicate potential low quality cells and multiplets. Cells with high percentage of mitochondrial genes indicates dead or dying cells that may not be of interest


```{r, results='asis', fig.width=5, fig.height=7, out.width="70%"}

cat("### RNA {-}\n\n")
cat(paste0("![](", file.path(seuratdir, "PreFilter_VlnPlot_RNA.png)")), "\n")
cat("\n")
```

```{r, results='asis', fig.width=5, fig.height=7, out.width="70%", eval=file.exists(file.path(seuratdir, "PreFilter_VlnPlot_ADT.png"))}

cat("### ADT {-}\n\n")
cat(paste0("![](", file.path(seuratdir, "PreFilter_VlnPlot_ADT.png)")), "\n")
cat("\n")
```

```{r, results='asis', fig.width=5, fig.height=7, out.width="70%", eval=file.exists(file.path(seuratdir, "PreFilter_VlnPlot_HTO.png"))}

cat("### HTO {-}\n\n")
cat(paste0("![](", file.path(seuratdir, "PreFilter_VlnPlot_HTO.png)")), "\n")
cat("\n")
```

## Pre-Filter Gene Plots

```{r, results='asis', fig.width=5, fig.height=7, out.width="70%"}

cat(paste0("![](", file.path(seuratdir, "PreFilter_Gene_Plot.png)")), "\n")
cat("\n")

```

## Pre-Filter UMAP Plot {.tabset}

The following are the UMAP plots for the sample that was generated prior to filtering cells.

```{r, results='asis', fig.width=5, fig.height=7, out.width="70%"}
if (file.exists(file.path(seuratdir, 'PreFilter_UMAP_RNA.png'))) {
  cat("### Clustering{-}\n\n")
  cat(paste0("![](", file.path(seuratdir, "PreFilter_UMAP_RNA.png)")), "\n")
}
cat("\n\n")
if(file.exists(file.path(seuratdir, 'PreFilter_UMAP_RNA_Filter.png'))) {
  cat("### Filtered Cells{-}\n\n")
  cat(paste0("![](", file.path(seuratdir, "PreFilter_UMAP_RNA_Filter.png)")), "\n")
}
cat("\n\n")

```

## Cell Filter Thresholds

```{r Cell Filter, warning=FALSE, message=FALSE, results='asis', fig.width=5, fig.height=7, out.width="70%"}
if (params$defaultfilter) {
  cat("Adaptive thresholds based on the feature distributions were used to filter out potential low quality cells and multiplets. The values were log-transformed and a threshold of 3 median absolute deviations (MADs) above and below the median was calculated. Only an upper threshold was calculated for mitochondrial percentage.^1^\n\n")
}else {
  cat("Thresholds were used to filter out potential low quality cells and multiplets.\n\n")
}
cat("The following are the thresholds that were set for each set of statistics.\n\n")

thresh <- read.csv(file.path(seuratdir, 'cell_filter_info.csv'))

printed <- c(apply(expand.grid(c("nFeature_RNA", "nCount_RNA", "nCount_ADT", "nCount_HTO", "percent.mito"), c("low", "high")), 1, paste, collapse='_'), "numCellsRemove", "pctCellsRemove")

get_value <- function(thresh, value) {
  if (is.null(thresh[[value]])) {NA} else{sprintf("%.0f", thresh[[value]])}
}
cat("GEX Gene Count Thresholds:", paste0("(", get_value(thresh, "nFeature_RNA_low"), ", ", get_value(thresh, "nFeature_RNA_high"), ")"), "\n\n")

cat("GEX Read Count Thresholds:", paste0("(", get_value(thresh, "nCount_RNA_low"), ", ", get_value(thresh, "nCount_RNA_high"), ")"), "\n\n")

cat("ADT Read Count Thresholds:", paste0("(", get_value(thresh, "nCount_ADT_low"), ", ", get_value(thresh, "nCount_ADT_high"), ")"), "\n\n")

cat("HTO Read Count Thresholds:", paste0("(", get_value(thresh, "nCount_HTO_low"), ", ", get_value(thresh, "nCount_HTO_high"), ")"), "\n\n")

cat("Mitochondrial Percentage Thresholds:", paste0("(", get_value(thresh, "percent.mito_low"), ", ", get_value(thresh, "percent.mito_high"), ")"), "\n\n")

cat("Total cells filtered:", get_value(thresh, "numCellsRemove"), "\n\n")

cat("Percentage cells filtered:", get_value(thresh, "pctCellsRemove"), "\n\n")

if (length(setdiff(names(thresh), printed) > 0)) {
  cat("The following additional thresholds were also used: \n\n")
  for (value in setdiff(names(thresh), printed)) {
    cat(paste0(value, ": ", thresh[[value]], "\n\n"))
    cat("\n\n")
  }
}
```


## Post-Filter Violin Plots {.tabset .tabset-pills}

The following are the resulting violin plots generated after filtering the cells.

```{r, results='asis', fig.width=5, fig.height=7, out.width="70%"}
cat("### RNA {-}\n\n")
cat(paste0("![](", file.path(seuratdir, 'PostFilter_VlnPlot_RNA.png)')), "\n")

```

```{r, results='asis', fig.width=5, fig.height=7, out.width="70%", eval=file.exists(file.path(seuratdir, "PostFilter_VlnPlot_ADT.png"))}

cat("### ADT {-}\n\n")
cat(paste0("![](", file.path(seuratdir, "PostFilter_VlnPlot_ADT.png)")), "\n")
cat("\n")
```

```{r, results='asis', fig.width=5, fig.height=7, out.width="70%", eval=file.exists(file.path(seuratdir, "PostFilter_VlnPlot_HTO.png"))}

cat("### HTO {-}\n\n")
cat(paste0("![](", file.path(seuratdir, "PostFilter_VlnPlot_HTO.png)")), "\n")
cat("\n")
```

## Post-Filter Gene Plots

```{r, results='asis', fig.width=5, fig.height=7, out.width="70%"}

cat(paste0("![](", file.path(seuratdir, "PostFilter_Gene_Plot.png)")), "\n")
cat("\n")

```

## Elbow Plot

```{r, results='asis', fig.width=5, fig.height=7, out.width="70%"}

cat(paste0("![](", file.path(seuratdir, "ElbowPlot.png)")), "\n")
cat("\n")

```

```{r, results='asis', eval=length(Sys.glob(file.path(seuratdir, 'HTO_*_count.csv'))) > 0}
cat("## Hashtag Demultiplexing\n\n")

filename <- Sys.glob(file.path(seuratdir, 'HTO_*_count.csv'))[1] 

   if (grep('hash.ID', filename) %>% length > 0) {
      cat("The function HTODemux was used to perform sample demultiplexing using a positive quantile threshold of", read.csv(file.path(seuratdir, "hto_threshold.csv"))[[1]], "\n\n")
   } else{
      cat("The function MULTIseqDemux was used to perform sample demultiplexing using a positive quantile threshold of", read.csv(file.path(seuratdir, "hto_threshold.csv"))[[1]], "\n\n")
   }

data <- read.csv(filename)
colnames(data) <- c('Hashtag', 'Count')
kable(data)

```

## Post-Filter UMAP Plot 

```{r, results='asis', fig.width=5, fig.height=7, out.width="70%", eval=length(Sys.glob(file.path(seuratdir, "UMAP_HTO_*"))) > 0}
cat("### Hashtag {.tabset .tabset-pills}\n\n")
for (filename in Sys.glob(file.path(seuratdir, "UMAP_HTO_*-split.png"))) {
  cat("#### Combined{-}\n\n")
  cat(paste0("![](", filename %>% gsub(pattern="-split", replacement=""), ')'), "\n\n")
  cat("#### Split{-}\n\n")
  cat(paste0("![](", filename, ')'), "\n\n")

}
```

### Clusters {.tabset}

The following are the UMAP plots for the sample that was generated after filtering out the potential low quality and multiplet cells.

```{r, results='asis', fig.width=5, fig.height=7, out.width="70%"}
files <- Sys.glob(file.path(seuratdir, "UMAP_RNA_res*.png"))
sil_files <- Sys.glob(file.path(seuratdir, "SilhouetteResult_res*.csv"))
resolutions <- sapply(files, function(x) (x %>% basename %>% tools::file_path_sans_ext() %>% strsplit(split='res.'))[[1]][2]) %>% str_sort(numeric=T)
sil_mean <- sapply(resolutions, function(resolution) {filename <- grep(paste0('res.', resolution, '.csv'), sil_files, value=T); if(length(filename) > 0) { data <- read.csv(filename); mean(data$sil_width)} else{-Inf}})

if (max(sil_mean) > -Inf) {
  cat("The average silhouette score and a silhouette plot was calculated for each cluster resolution. This was done by calculating the score for each cell using the euclidean distance generated based on the cell's principal components to obtain an independent measure of the clustering results. The silhouette score ranges from -1 to 1, where: \n\n* 1 means that clusters are far apart and data points close to each other are assigned the same cluster label\n\n* 0 means that the distance between clusters are not signifcant and data points are close to more than one cluster\n\n* -1 means that clusters may be incorrectly assigned and data points are closer to other clusters than other data points in its own cluster\n\nThe resolution with the highest detected average silhouette score is resolution", sil_mean %>% which.max %>% names, "\n\n")
}

for (resolution in resolutions) {
  filename <- grep(paste0('res.', resolution, '.png'), files, value=T)
  cat("#### Resolution", resolution, "{.tabset .tabset-fade .tabset-pills -}\n\n")
  #  cat("#### UMAP {-}\n\n")
  cat(paste0("![](", filename, ')'), "\n\n")
  sil_filename <- filename %>% gsub(pattern = 'UMAP_RNA', replacement = "SilhouettePlot") %>% gsub(pattern=".png", replacement=".pdf")
  if (file.exists(sil_filename)) {
    if (file.exists(sil_filename %>% gsub(pattern='Plot', replacement='Result') %>% gsub(pattern='pdf', replacement='csv'))) {
      data <- read.csv(sil_filename %>% gsub(pattern='Plot', replacement='Result') %>% gsub(pattern='pdf', replacement='csv'))
      cat("Average silhouette score:", mean(data$sil_width) %>% round(digits=3), "\n\n")
    } else{
      cat("Average silhouette score:", "NA", "\n\n")
    }
    print(xfun::embed_file(sil_filename, text = paste0("Download Silhouette Plot Resolution ", resolution)))#
    cat("The silhouette plot itself labels the clusters starting from the number 1. The clusters are still in the order that is listed in the UMAP figure, it is just labeled from 1 onward.\n\n")
    cat("\n\n")
    #    cat("#### Silhouette Plot {-}\n\n")
    #    cat(paste0("![](", 'data:application/pdf;base64,', base64encode(sil_filename), "){width=100%}"), "\n\n")
  }
  #cat(paste0("![](", 'data:application/pdf;base64,', base64encode('/gpfs/gsfs12/users/NHLBI_IDSS/dev/cell-seek-test/cellranger8/testdata/custom_ref/NCBR339_8_gex/seurat/Tube_5__Sample_5_GEX_library/SilhouettePlot_res.0.2.pdf'), ")"), "\n\n")
}

```

## ADT Expression {.tabset .tabset-pills}
Ridge plots have been generated for each of the ADT included in the data. 

```{r, results='asis'}

myimages<-list.files(file.path(seuratdir, "ADT/"), pattern = ".png", full.names = TRUE)
for (filename in myimages) {
  if (length(myimages) > 1) {
    cat("###", tail(str_split(filename %>% tools::file_path_sans_ext(), pattern='_')[[1]], n=1), "{-}\n\n")
  }
  cat(paste0("![](", filename, ')'), "\n\n")
}
```

```{r, results='asis', eval=file.exists(file.path(seuratdir, "HTO_Ridge_Plot.png"))}
cat("## HTO Expression\n\n")
cat(paste0("![](", file.path(seuratdir, "HTO_Ridge_Plot.png"), ')'), "\n\n")
```

# Session Info

``` {r Session Info, comment='', results='markup', attr.output='style="max-height: 500px;"'}
cat(readLines(file.path(seuratdir, 'sessionInfo.txt')), sep="\n")
```

``` {r References, eval=params$defaultfilter}
cat("# References\n\n")
cat("1. Lun AT, McCarthy DJ, Marioni JC. A step-by-step workflow for low-level analysis of single-cell RNA-seq data with Bioconductor. F1000Res. 2016 Aug 31;5:2122. doi: 10.12688/f1000research.9501.2. PMID: 27909575; PMCID: PMC5112579.\n\n")
```

<div class="tocify-extend-page" data-unique="tocify-extend-page" style="height: 0;"></div>
