---
title: "`r params$project`"
subtitle: "Single Cell RNA-seq - QC"
date: "`r Sys.Date()`"
output: 
  html_document:
    lightbox: true
    toc: false
    toc_float: 
        collapsed: false
    toc_depth: 3
    fig_width: 8
    fig_height: 5
    number_sections: false
params:
  project: Project
  seuratdir: directory
  samples: samples
  cellfilter: cellfilter
  skipped: ""
---



```{r setup, include=FALSE}

# Setup directory ####

knitr::opts_chunk$set(echo = TRUE)
library(knitr, quietly = T)
library(kableExtra)
library(stringr)

plt <- function(sample, plot.name) {
  plot.loc <- file.path(maindir, sample, plot.name)
  return(try(plot.loc))
}

maindir <-params$seuratdir
plt1 <- "PreFilter_Gene_Plot.png"
plt2 <- "PostFilter_Gene_Plot.png"
plt3 <- "PreFilter_VlnPlot_RNA.png"
plt4 <- "PostFilter_VlnPlot_RNA.png"
plt5 <- "PreFilter_UMAP_RNA_Filter.png"
#plt8 <- "TopVariableGenes.png"
plt9 <- "UMAP_RNA_res.0.8.png"
plt_pre_adt <- "PreFilter_VlnPlot_ADT.png"
plt_post_adt <- "PostFilter_VlnPlot_ADT.png"
plt_pre_hto <- "PreFilter_VlnPlot_HTO.png"
plt_post_hto <- "PostFilter_VlnPlot_HTO.png"
plt_pre_mito <- "PreFilter_UMAP_percent.mito.png"
plt_post_mito <- "PostFilter_UMAP_percent.mito.png"
plt_pre_rpl <- "PreFilter_UMAP_percent.rpl.png"
plt_post_rpl <- "PostFilter_UMAP_percent.rpl.png"
plt_pre_rps <- "PreFilter_UMAP_percent.rps.png"
plt_post_rps <- "PostFilter_UMAP_percent.rps.png"
plt_pre_feature_rna <- "PreFilter_UMAP_nFeature_RNA.png" 
plt_post_feature_rna  <- "PostFilter_UMAP_nFeature_RNA.png"
plt_pre_count_rna <- "PreFilter_UMAP_nCount_RNA.png"
plt_post_count_rna <- "PostFilter_UMAP_nCount_RNA.png" 

sil_wildcard <- "SilhouetteResult_res*.csv"

sample_list <- params$samples
sample_list <- str_sort(sample_list, numeric=T)

```

# QC analysis  {.tabset .tabset-fade}

## RNA Pre-Filter QC Plots

```{r  echo=F ,results='asis'}

#colnames(data) <- c('Sample', 'Violin Plots', 'Scatter Plots', 'UMAP Plots')
data <- cbind(sample_list |> gsub(pattern='/', replacement=' - '), lapply(sample_list, function(x) paste0(file.path(maindir, x, plt3) %>% pander::pandoc.image.return(), '{height=260px, width=260px}')), lapply(sample_list, function(x) paste0(file.path(maindir, x, plt1) %>% pander::pandoc.image.return(), '{height=260px, width=780px}'))) %>% as.data.frame()
colnames(data) <- c('Sample', 'Violin Plots', 'Scatter Plots')
data %>% kable(format='html') %>% 
	column_spec(column=length(data) %>% seq(), border_left = "1px solid gray", border_right = "1px solid gray") %>% 
	row_spec(row=dim(data)[[1]] %>% seq(), hline_after=T, align='c', extra_css = "border-bottom: 1px solid gray") %>%
	row_spec(row=0,bold=TRUE, align='c', extra_css = "border: 1px solid gray")
```

```{r echo=F, results='asis', eval=length(Sys.glob(file.path(maindir, '*', plt_pre_adt))) > 0}
cat("## ADT Pre-Filter QC Plots\n\n")
data <- cbind(sample_list |> gsub(pattern='/', replacement=' - '), lapply(sample_list, function(x) paste0(file.path(maindir, x, plt_pre_adt) %>% pander::pandoc.image.return(), '{height=260px, width=260px}'))) %>% as.data.frame()
colnames(data) <- c('Sample', 'Violin Plots')
data %>% kable(format='html') %>% 
	column_spec(column=length(data) %>% seq(), border_left = "1px solid gray", border_right = "1px solid gray") %>% 
	row_spec(row=dim(data)[[1]] %>% seq(), hline_after=T, align='c', extra_css = "border-bottom: 1px solid gray") %>%
	row_spec(row=0,bold=TRUE, align='c', extra_css = "border: 1px solid gray")
```

```{r echo=F, results='asis', eval=length(Sys.glob(file.path(maindir, '*', plt_pre_hto))) > 0}
cat("## HTO Pre-Filter QC Plots\n\n")
data <- cbind(sample_list |> gsub(pattern='/', replacement=' - '), lapply(sample_list, function(x) paste0(file.path(maindir, x, plt_pre_hto) %>% pander::pandoc.image.return(), '{height=260px, width=260px}'))) %>% as.data.frame()
colnames(data) <- c('Sample', 'Violin Plots')
data %>% kable(format='html') %>% 
	column_spec(column=length(data) %>% seq(), border_left = "1px solid gray", border_right = "1px solid gray") %>% 
	row_spec(row=dim(data)[[1]] %>% seq(), hline_after=T, align='c', extra_css = "border-bottom: 1px solid gray") %>%
	row_spec(row=0,bold=TRUE, align='c', extra_css = "border: 1px solid gray")
```

## Pre-Filter UMAP

```{r  echo=F ,results='asis'}
data <- cbind(sample_list |> gsub(pattern='/', replacement=' - '), lapply(sample_list, function(x) paste0(file.path(maindir, x, plt5) %>% pander::pandoc.image.return(), '{height=512px, width=576px}')), lapply(sample_list, function(x) paste0(file.path(maindir, x, plt_pre_feature_rna) %>% pander::pandoc.image.return(), '{height=512px, width=576px}')), lapply(sample_list, function(x) paste0(file.path(maindir, x, plt_pre_count_rna) %>% pander::pandoc.image.return(), '{height=512px, width=576px}')), lapply(sample_list, function(x) paste0(file.path(maindir, x, plt_pre_mito) %>% pander::pandoc.image.return(), '{height=512px, width=576px}')), lapply(sample_list, function(x) paste0(file.path(maindir, x, plt_pre_rpl) %>% pander::pandoc.image.return(), '{height=512px, width=576px}')), lapply(sample_list, function(x) paste0(file.path(maindir, x, plt_pre_rps) %>% pander::pandoc.image.return(), '{height=512px, width=576px}'))) %>% as.data.frame()
colnames(data) <- c('Sample', 'Pre-Filter UMAP', 'Pre-Filter Number of Molecules', 'Pre-Filter Number of Genes', 'Pre-Filter Percent Mito', 'Pre-Filter Percent Ribosomal Protein (Large)', 'Pre-Filter Percent Ribosomal Protein (Small)')
data %>% kable(format='html') %>%
        column_spec(column=length(data) %>% seq(), border_left = "1px solid gray", border_right = "1px solid gray") %>%
        row_spec(row=dim(data)[[1]] %>% seq(), hline_after=T, align='c', extra_css = "border-bottom: 1px solid gray") %>%
        row_spec(row=0,bold=TRUE, align='c', extra_css = "border: 1px solid gray")

```

## Cell Filter Thresholds

```{r  echo=F ,results='asis'}
data <- read.csv(params$cellfilter)
colnames(data) <- gsub('_low', ' - Lower', colnames(data))
colnames(data) <- gsub('_high', ' - Upper', colnames(data))
colnames(data) <- gsub('initialCellCount', 'Initial Cell Count', colnames(data))
colnames(data) <- gsub('finalCellCount', 'Final Cell Count', colnames(data))
colnames(data) <- gsub('numCellsRemove', 'Number of Cells Removed', colnames(data))
colnames(data) <- gsub('pctCellsRemove', 'Percentage of Cells Removed', colnames(data))
kable(data, format='html') %>%
        column_spec(column=length(data) %>% seq(), border_left = "1px solid gray", border_right = "1px solid gray") %>%
        row_spec(row=dim(data)[[1]] %>% seq(), hline_after=T, align='c', extra_css = "border-bottom: 1px solid gray") %>%
        row_spec(row=0,bold=TRUE, align='c', extra_css = "border: 1px solid gray")
```

## RNA Post-Filter QC Plots

```{r  echo=F ,results='asis'}
data <- cbind(sample_list |> gsub(pattern='/', replacement=' - '), lapply(sample_list, function(x) paste0(file.path(maindir, x, plt4) %>% pander::pandoc.image.return(), '{height=260px, width=260px}')), lapply(sample_list, function(x) paste0(file.path(maindir, x, plt2) %>% pander::pandoc.image.return(), '{height=260px, width=780px}'))) %>% as.data.frame()
colnames(data) <- c('Sample', 'Violin Plots', 'Scatter Plots')
data %>% kable(format='html') %>%
        column_spec(column=length(data) %>% seq(), border_left = "1px solid gray", border_right = "1px solid gray") %>%
        row_spec(row=dim(data)[[1]] %>% seq(), hline_after=T, align='c', extra_css = "border-bottom: 1px solid gray") %>%
        row_spec(row=0,bold=TRUE, align='c', extra_css = "border: 1px solid gray")

```

```{r echo=F, results='asis', eval=length(Sys.glob(file.path(maindir, '*', plt_pre_adt))) > 0}
cat("## ADT Post-Filter QC Plots\n\n")
data <- cbind(sample_list |> gsub(pattern='/', replacement=' - '), lapply(sample_list, function(x) paste0(file.path(maindir, x, plt_post_adt) %>% pander::pandoc.image.return(), '{height=260px, width=260px}'))) %>% as.data.frame()
colnames(data) <- c('Sample', 'Violin Plots')
data %>% kable(format='html') %>% 
	column_spec(column=length(data) %>% seq(), border_left = "1px solid gray", border_right = "1px solid gray") %>% 
	row_spec(row=dim(data)[[1]] %>% seq(), hline_after=T, align='c', extra_css = "border-bottom: 1px solid gray") %>%
	row_spec(row=0,bold=TRUE, align='c', extra_css = "border: 1px solid gray")
```

```{r echo=F, results='asis', eval=length(Sys.glob(file.path(maindir, '*', plt_pre_hto))) > 0}
cat("## HTO Post-Filter QC Plots\n\n")
data <- cbind(sample_list |> gsub(pattern='/', replacement=' - '), lapply(sample_list, function(x) paste0(file.path(maindir, x, plt_post_hto) %>% pander::pandoc.image.return(), '{height=260px, width=260px}'))) %>% as.data.frame()
colnames(data) <- c('Sample', 'Violin Plots')
data %>% kable(format='html') %>% 
	column_spec(column=length(data) %>% seq(), border_left = "1px solid gray", border_right = "1px solid gray") %>% 
	row_spec(row=dim(data)[[1]] %>% seq(), hline_after=T, align='c', extra_css = "border-bottom: 1px solid gray") %>%
	row_spec(row=0,bold=TRUE, align='c', extra_css = "border: 1px solid gray")
```


## Post-Filter UMAP

```{r  echo=F ,results='asis'}
#Setting default resolution to 0.8
resolutions <- sapply(sample_list, function(x) 0.8)

#Trying to extract the resolution with the highest silhouette score for each sample. Error messages are suppressed to prevent them from showing up on the report if this fails.
try(expr={

  #Going through each sample to get the mean silhouette score for each resolution where the silhouette results are found
  sil_max <- sapply(sample_list, function(x) {
    sil_files <- Sys.glob(file.path(maindir, x, sil_wildcard))
    which.max(sapply(sil_files, function(filename) {data <- read.csv(filename); mean(data$sil_width)}))
  } )

  #Extracting resolution from file name and sample name for each resolution
  resolutions <- (sil_max %>% names %>% basename %>% tools::file_path_sans_ext() %>% str_split_fixed(pattern='_res.', n=Inf))[,2]
  names(resolutions) <- sample_list
  cat("The UMAP plot for the resolution with the highest silhouette score was selected for each sample.\n\n")
}, silent=TRUE)

#Create data table while replacing the 0.8 from the original plt9 file name and replacing it with resolution from resolution list
data <- cbind(sample_list |> gsub(pattern='/', replacement=' - '), lapply(sample_list, function(x) paste0(file.path(maindir, x, plt9 %>% gsub(pattern='0.8', replacement=resolutions[x])) %>% pander::pandoc.image.return(), '{height=512px, width=576px}')), lapply(sample_list, function(x) paste0(file.path(maindir, x, plt_post_feature_rna) %>% pander::pandoc.image.return(), '{height=512px, width=576px}')), lapply(sample_list, function(x) paste0(file.path(maindir, x, plt_post_count_rna) %>% pander::pandoc.image.return(), '{height=512px, width=576px}')), lapply(sample_list, function(x) paste0(file.path(maindir, x, plt_post_mito) %>% pander::pandoc.image.return(), '{height=512px, width=576px}')), lapply(sample_list, function(x) paste0(file.path(maindir, x, plt_post_rpl) %>% pander::pandoc.image.return(), '{height=512px, width=576px}')), lapply(sample_list, function(x) paste0(file.path(maindir, x, plt_post_rps) %>% pander::pandoc.image.return(), '{height=512px, width=576px}'))) %>% as.data.frame()
colnames(data) <- c('Sample', 'Seurat Clusters', 'Post-Filter Number of Molecules', 'Post-Filter Number of Genes', 'Post-Filter Percent Mito', 'Post-Filter Percent Ribosomal Protein (Large)', 'Post-Filter Percent Ribosomal Protein (Small)')
data %>% kable(format='html') %>%
        column_spec(column=length(data) %>% seq(), border_left = "1px solid gray", border_right = "1px solid gray") %>%
        row_spec(row=dim(data)[[1]] %>% seq(), hline_after=T, align='c', extra_css = "border-bottom: 1px solid gray") %>%
        row_spec(row=0,bold=TRUE, align='c', extra_css = "border: 1px solid gray")

```

```{r  echo=F ,results='asis', eval=length(setdiff(params$skipped, '')) > 0}
cat("## Skipped Samples\n\n")
cat("The following are a list of samples and tags that did not return enough cells after Cell Ranger to proceed with downstream analysis. Please check to ensure that the this list does not contain any samples where cells are expected.\n\n")
cat(sprintf("* %s\n\n", paste(str_sort(params$skipped, numeric=T), collapse="\n* ")))
```

