<!doctype html><html lang=en class=no-js> <head><meta charset=utf-8><meta name=viewport content="width=device-width,initial-scale=1"><meta name=description content="An awesome OpenOmics cell-seek pipeline"><meta name=author content="Skyler Kuhn"><link rel=prev href=../..><link href=../genome/ rel=next><link rel=icon href=../../assets/favicon/favicon.ico><meta name=generator content="mkdocs-1.6.1, mkdocs-material-9.7.1"><title>cell-seek run - cell-seek</title><link rel=stylesheet href=../../assets/stylesheets/main.484c7ddc.min.css><link rel=stylesheet href=../../assets/stylesheets/palette.ab4e12ef.min.css><link rel=preconnect href=https://fonts.gstatic.com crossorigin><link rel=stylesheet href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback"><style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style><link rel=stylesheet href=../../css/extra.css><script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script></head> <body dir=ltr data-md-color-scheme=default data-md-color-primary=indigo data-md-color-accent=indigo> <input class=md-toggle data-md-toggle=drawer type=checkbox id=__drawer autocomplete=off> <input class=md-toggle data-md-toggle=search type=checkbox id=__search autocomplete=off> <label class=md-overlay for=__drawer></label> <div data-md-component=skip> <a href=#cell-seek-run class=md-skip> Skip to content </a> </div> <div data-md-component=announce> </div> <div data-md-color-scheme=default data-md-component=outdated hidden> </div> <header class=md-header data-md-component=header> <nav class="md-header__inner md-grid" aria-label=Header> <a href=../.. title=cell-seek class="md-header__button md-logo" aria-label=cell-seek data-md-component=logo> <img src=../../assets/icons/doc-book.svg alt=logo> </a> <label class="md-header__button md-icon" for=__drawer> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg> </label> <div class=md-header__title data-md-component=header-title> <div class=md-header__ellipsis> <div class=md-header__topic> <span class=md-ellipsis> cell-seek </span> </div> <div class=md-header__topic data-md-component=header-topic> <span class=md-ellipsis> cell-seek run </span> </div> </div> </div> <form class=md-header__option data-md-component=palette> <input class=md-option data-md-color-media data-md-color-scheme=default data-md-color-primary=indigo data-md-color-accent=indigo aria-label="Switch to dark mode" type=radio name=__palette id=__palette_0> <label class="md-header__button md-icon" title="Switch to dark mode" for=__palette_1 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 6a6 6 0 0 1 6 6c0 2.22-1.21 4.16-3 5.2V19a1 1 0 0 1-1 1h-4a1 1 0 0 1-1-1v-1.8c-1.79-1.04-3-2.98-3-5.2a6 6 0 0 1 6-6m2 15v1a1 1 0 0 1-1 1h-2a1 1 0 0 1-1-1v-1zm6-10h3v2h-3zM1 11h3v2H1zM13 1v3h-2V1zM4.92 3.5l2.13 2.14-1.42 1.41L3.5 4.93zm12.03 2.13 2.12-2.13 1.43 1.43-2.13 2.12z"/></svg> </label> <input class=md-option data-md-color-media data-md-color-scheme=slate data-md-color-primary=indigo data-md-color-accent=indigo aria-label="Switch to light mode" type=radio name=__palette id=__palette_1> <label class="md-header__button md-icon" title="Switch to light mode" for=__palette_0 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="m17.75 4.09-2.53 1.94.91 3.06-2.63-1.81-2.63 1.81.91-3.06-2.53-1.94L12.44 4l1.06-3 1.06 3zm3.5 6.91-1.64 1.25.59 1.98-1.7-1.17-1.7 1.17.59-1.98L15.75 11l2.06-.05L18.5 9l.69 1.95zm-2.28 4.95c.83-.08 1.72 1.1 1.19 1.85-.32.45-.66.87-1.08 1.27C15.17 23 8.84 23 4.94 19.07c-3.91-3.9-3.91-10.24 0-14.14.4-.4.82-.76 1.27-1.08.75-.53 1.93.36 1.85 1.19-.27 2.86.69 5.83 2.89 8.02a9.96 9.96 0 0 0 8.02 2.89m-1.64 2.02a12.08 12.08 0 0 1-7.8-3.47c-2.17-2.19-3.33-5-3.49-7.82-2.81 3.14-2.7 7.96.31 10.98 3.02 3.01 7.84 3.12 10.98.31"/></svg> </label> </form> <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script> <label class="md-header__button md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg> </label> <div class=md-search data-md-component=search role=dialog> <label class=md-search__overlay for=__search></label> <div class=md-search__inner role=search> <form class=md-search__form name=search> <input type=text class=md-search__input name=query aria-label=Search placeholder=Search autocapitalize=off autocorrect=off autocomplete=off spellcheck=false data-md-component=search-query required> <label class="md-search__icon md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg> </label> <nav class=md-search__options aria-label=Search> <button type=reset class="md-search__icon md-icon" title=Clear aria-label=Clear tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg> </button> </nav> </form> <div class=md-search__output> <div class=md-search__scrollwrap tabindex=0 data-md-scrollfix> <div class=md-search-result data-md-component=search-result> <div class=md-search-result__meta> Initializing search </div> <ol class=md-search-result__list role=presentation></ol> </div> </div> </div> </div> </div> <div class=md-header__source> <a href=https://github.com/OpenOmics/cell-seek title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg> </div> <div class=md-source__repository> OpenOmics/cell-seek </div> </a> </div> </nav> </header> <div class=md-container data-md-component=container> <nav class=md-tabs aria-label=Tabs data-md-component=tabs> <div class=md-grid> <ul class=md-tabs__list> <li class=md-tabs__item> <a href=../.. class=md-tabs__link> About </a> </li> <li class="md-tabs__item md-tabs__item--active"> <a href=./ class=md-tabs__link> Commands </a> </li> <li class=md-tabs__item> <a href=../../faq/questions/ class=md-tabs__link> FAQ </a> </li> <li class=md-tabs__item> <a href=../../license/ class=md-tabs__link> License </a> </li> </ul> </div> </nav> <main class=md-main data-md-component=main> <div class="md-main__inner md-grid"> <div class="md-sidebar md-sidebar--primary" data-md-component=sidebar data-md-type=navigation> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--primary md-nav--lifted md-nav--integrated" aria-label=Navigation data-md-level=0> <label class=md-nav__title for=__drawer> <a href=../.. title=cell-seek class="md-nav__button md-logo" aria-label=cell-seek data-md-component=logo> <img src=../../assets/icons/doc-book.svg alt=logo> </a> cell-seek </label> <div class=md-nav__source> <a href=https://github.com/OpenOmics/cell-seek title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg> </div> <div class=md-source__repository> OpenOmics/cell-seek </div> </a> </div> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../.. class=md-nav__link> <span class=md-ellipsis> About </span> </a> </li> <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2 checked> <label class=md-nav__link for=__nav_2 id=__nav_2_label tabindex> <span class=md-ellipsis> Commands </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_2_label aria-expanded=true> <label class=md-nav__title for=__nav_2> <span class="md-nav__icon md-icon"></span> Commands </label> <ul class=md-nav__list data-md-scrollfix> <li class="md-nav__item md-nav__item--active"> <input class="md-nav__toggle md-toggle" type=checkbox id=__toc> <label class="md-nav__link md-nav__link--active" for=__toc> <span class=md-ellipsis> cell-seek run </span> <span class="md-nav__icon md-icon"></span> </label> <a href=./ class="md-nav__link md-nav__link--active"> <span class=md-ellipsis> cell-seek run </span> </a> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#1-about class=md-nav__link> <span class=md-ellipsis> 1. About </span> </a> </li> <li class=md-nav__item> <a href=#2-synopsis class=md-nav__link> <span class=md-ellipsis> 2. Synopsis </span> </a> <nav class=md-nav aria-label="2. Synopsis"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#21-gex class=md-nav__link> <span class=md-ellipsis> 2.1 GEX </span> </a> <nav class=md-nav aria-label="2.1 GEX"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#211-required-arguments class=md-nav__link> <span class=md-ellipsis> 2.1.1 Required Arguments </span> </a> </li> <li class=md-nav__item> <a href=#212-analysis-options class=md-nav__link> <span class=md-ellipsis> 2.1.2 Analysis Options </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#22-vdj class=md-nav__link> <span class=md-ellipsis> 2.2 VDJ </span> </a> <nav class=md-nav aria-label="2.2 VDJ"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#221-required-arguments class=md-nav__link> <span class=md-ellipsis> 2.2.1 Required Arguments </span> </a> </li> <li class=md-nav__item> <a href=#222-analysis-options class=md-nav__link> <span class=md-ellipsis> 2.2.2 Analysis Options </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#23-cite class=md-nav__link> <span class=md-ellipsis> 2.3 CITE </span> </a> <nav class=md-nav aria-label="2.3 CITE"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#231-required-arguments class=md-nav__link> <span class=md-ellipsis> 2.3.1 Required Arguments </span> </a> </li> <li class=md-nav__item> <a href=#232-conditionally-required-arguments class=md-nav__link> <span class=md-ellipsis> 2.3.2 Conditionally Required Arguments </span> </a> </li> <li class=md-nav__item> <a href=#233-analysis-options class=md-nav__link> <span class=md-ellipsis> 2.3.3 Analysis Options </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#24-multi class=md-nav__link> <span class=md-ellipsis> 2.4 MULTI </span> </a> <nav class=md-nav aria-label="2.4 MULTI"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#241-required-arguments class=md-nav__link> <span class=md-ellipsis> 2.4.1 Required Arguments </span> </a> </li> <li class=md-nav__item> <a href=#242-conditionally-required-arguments class=md-nav__link> <span class=md-ellipsis> 2.4.2 Conditionally Required Arguments </span> </a> </li> <li class=md-nav__item> <a href=#243-analysis-options class=md-nav__link> <span class=md-ellipsis> 2.4.3 Analysis Options </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#25-atac class=md-nav__link> <span class=md-ellipsis> 2.5 ATAC </span> </a> <nav class=md-nav aria-label="2.5 ATAC"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#251-required-arguments class=md-nav__link> <span class=md-ellipsis> 2.5.1 Required Arguments </span> </a> </li> <li class=md-nav__item> <a href=#252-analysis-options class=md-nav__link> <span class=md-ellipsis> 2.5.2 Analysis Options </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#26-multiome class=md-nav__link> <span class=md-ellipsis> 2.6 Multiome </span> </a> <nav class=md-nav aria-label="2.6 Multiome"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#261-required-arguments class=md-nav__link> <span class=md-ellipsis> 2.6.1 Required Arguments </span> </a> </li> <li class=md-nav__item> <a href=#262-conditionally-required-arguments class=md-nav__link> <span class=md-ellipsis> 2.6.2 Conditionally Required Arguments </span> </a> </li> <li class=md-nav__item> <a href=#263-analysis-options class=md-nav__link> <span class=md-ellipsis> 2.6.3 Analysis Options </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#27-orchestration-options class=md-nav__link> <span class=md-ellipsis> 2.7 Orchestration Options </span> </a> </li> <li class=md-nav__item> <a href=#28-miscellaneous-options class=md-nav__link> <span class=md-ellipsis> 2.8 Miscellaneous Options </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#3-example class=md-nav__link> <span class=md-ellipsis> 3. Example </span> </a> <nav class=md-nav aria-label="3. Example"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#31-gex-example class=md-nav__link> <span class=md-ellipsis> 3.1 GEX Example </span> </a> </li> <li class=md-nav__item> <a href=#32-vdj-example class=md-nav__link> <span class=md-ellipsis> 3.2 VDJ Example </span> </a> </li> <li class=md-nav__item> <a href=#33-cite-example class=md-nav__link> <span class=md-ellipsis> 3.3 CITE Example </span> </a> </li> <li class=md-nav__item> <a href=#34-multi-example class=md-nav__link> <span class=md-ellipsis> 3.4 MULTI Example </span> </a> <nav class=md-nav aria-label="3.4 MULTI Example"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#341-gex-and-vdj class=md-nav__link> <span class=md-ellipsis> 3.4.1 GEX and VDJ </span> </a> </li> <li class=md-nav__item> <a href=#342-including-hto class=md-nav__link> <span class=md-ellipsis> 3.4.2 Including HTO </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#35-atac-example class=md-nav__link> <span class=md-ellipsis> 3.5 ATAC Example </span> </a> </li> <li class=md-nav__item> <a href=#36-multiome-example class=md-nav__link> <span class=md-ellipsis> 3.6 Multiome Example </span> </a> </li> </ul> </nav> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=../genome/ class=md-nav__link> <span class=md-ellipsis> cell-seek genome </span> </a> </li> <li class=md-nav__item> <a href=../unlock/ class=md-nav__link> <span class=md-ellipsis> cell-seek unlock </span> </a> </li> <li class=md-nav__item> <a href=../shinycell/ class=md-nav__link> <span class=md-ellipsis> cell-seek shinycell </span> </a> </li> <li class=md-nav__item> <a href=../cache/ class=md-nav__link> <span class=md-ellipsis> cell-seek cache </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_3> <label class=md-nav__link for=__nav_3 id=__nav_3_label tabindex=0> <span class=md-ellipsis> FAQ </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_3_label aria-expanded=false> <label class=md-nav__title for=__nav_3> <span class="md-nav__icon md-icon"></span> FAQ </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../faq/questions/ class=md-nav__link> <span class=md-ellipsis> General Questions </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=../../license/ class=md-nav__link> <span class=md-ellipsis> License </span> </a> </li> </ul> </nav> </div> </div> </div> <div class=md-content data-md-component=content> <article class="md-content__inner md-typeset"> <h1 id=cell-seek-run><code>cell-seek <b>run</b></code><a class=headerlink href=#cell-seek-run title="Permanent link">&para;</a></h1> <h2 id=1-about>1. About<a class=headerlink href=#1-about title="Permanent link">&para;</a></h2> <p>The <code>cell-seek</code> executable is composed of several inter-related sub commands. Please see <code>cell-seek -h</code> for all available options.</p> <p>This part of the documentation describes options and concepts for <code>cell-seek <b>run</b></code> sub command in more detail. The following page lists the options applicable to each of the different pipelines available to select via the <code><b>--pipeline</b></code> flag. With minimal configuration, the <strong><code>run</code></strong> sub command enables you to start running cell-seek pipeline.</p> <p>Setting up the cell-seek pipeline is fast and easy! In its most basic form, <code>cell-seek <b>run</b></code> only has <em>four required inputs</em>.</p> <h2 id=2-synopsis>2. Synopsis<a class=headerlink href=#2-synopsis title="Permanent link">&para;</a></h2> <div class=highlight><pre><span></span><code>$ cell-seek run [--help] \
      [--dry-run] [--job-name JOB_NAME] [--mode {{slurm,local}}] \
      [--sif-cache SIF_CACHE] [--singularity-cache SINGULARITY_CACHE] \
      [--silent] [--threads THREADS] [--tmp-dir TMP_DIR] \
      [--cellranger {8.0.0, ...} \
      [--aggregate {{mapped, none}}] [--exclude-introns] \
      [--libraries LIBRARIES] [--features FEATURES] \
      [--filter FILTER] [--metadata METADATA] [--create-bam] \
      [--rename RENAME] \
      --input INPUT [INPUT ...] \
      --output OUTPUT \
      --pipeline {gex, ...} \
      --genome {hg38, ...}
</code></pre></div> <p>The synopsis for each command shows its arguments and their usage. Optional arguments are shown in square brackets.</p> <p>A user <strong>must</strong> provide a list of FastQ (globbing is supported) to analyze via <code>--input</code> argument, an output directory to store results via <code>--output</code> argument, the version of the pipeline to run via <code>--pipeline</code> argument, and the reference genome to use via <code>--genome</code> argument.</p> <p>Use you can always use the <code>-h</code> option for information on a specific command.</p> <p>The following is a breakdown of the required and optional arguments for each of the versions of the pipeline.</p> <h3 id=21-gex>2.1 GEX<a class=headerlink href=#21-gex title="Permanent link">&para;</a></h3> <h4 id=211-required-arguments>2.1.1 Required Arguments<a class=headerlink href=#211-required-arguments title="Permanent link">&para;</a></h4> <p>Each of the following arguments are required. Failure to provide a required argument will result in a non-zero exit-code.</p> <p><code>--input INPUT [INPUT ...]</code> </p> <blockquote> <p><strong>Input FastQ file(s) or Cell Ranger folder(s).</strong><br> <em>type: file(s) or folder(s)</em> </p> <p>FastQ Input: One or more FastQ files can be provided. The pipeline does NOT support single-end data. From the command-line, each input file should separated by a space. Multiple input FastQ files per sample can be provided. Globbing is supported! This makes selecting FastQ files easy. Input FastQ files should always be gzipp-ed.</p> <p><strong><em>Example:</em></strong> <code>--input .tests/*.R?.fastq.gz</code></p> <p>Cell Ranger Input: Cell Ranger output folders can be provided. It is expected that the outs folder is contained within the Cell Ranger output folders, and keep the normal output folder structure. Globbing is supported!</p> <p><strong><em>Example:</em></strong> <code>--input .tests/*/</code></p> </blockquote> <hr> <p><code>--output OUTPUT</code></p> <blockquote> <p><strong>Path to an output directory.</strong> <br> <em>type: path</em></p> <p>This location is where the pipeline will create all of its output files, also known as the pipeline's working directory. If the provided output directory does not exist, it will be created automatically.</p> <p><strong><em>Example:</em></strong> <code>--output /data/$USER/cell-seek_out</code></p> </blockquote> <hr> <p><code>--pipeline gex</code></p> <blockquote> <p><strong>The pipeline to run.</strong> <br> <em>type: string</em></p> <p>This option selects the pipeline to run. The documentation provided is based on choosing the option for gene expression (GEX).</p> <p><strong><em>Example:</em></strong> <code>--pipeline gex</code></p> </blockquote> <hr> <p><code>--genome {hg38, mm10, hg2024, mm2024, custom.json}</code></p> <blockquote> <p><strong>Reference genome.</strong> <br> <em>type: string</em></p> <p>This option defines the reference genome of the samples. cell-seek does comes bundled with prebuilt reference files for human and mouse samples, The options hg38 or mm10 would select the 2020 release of the reference. The options hg2024 or mm2024 would select the 2024 release of the reference. More information about the officially released references can be found on the <a href=https://www.10xgenomics.com/support/software/cell-ranger/latest/release-notes/cr-reference-release-notes>10x Genomics website</a>.</p> <p>A custom reference genome can also be provided via a json file. Additional information for creating this json file can be found in <a href=../genome/ ><code>cell-seek <b>genome</b></code></a>.</p> <p>For prebuilt references please select one of the following options: hg38, mm10, hg2024, mm2024</p> <p><strong><em>Example:</em></strong> <code>--genome hg2024</code></p> </blockquote> <hr> <p><code>--cellranger {7.1.0, 7.2.0, 8.0.0, 9.0.0, 10.0.0}</code></p> <blockquote> <p><strong>The version of Cell Ranger to run.</strong> <br> <em>type: string</em></p> <p>This option specifies which version of Cell Ranger to use when running GEX, VDJ, CITE, or MULTI pipelines. Please select one of the following options: 7.1.0, 7.2.0, 8.0.0, 9.0.0, 10.0.0</p> <p><strong><em>Example:</em></strong> <code>--cellranger 7.1.0</code></p> </blockquote> <h4 id=212-analysis-options>2.1.2 Analysis Options<a class=headerlink href=#212-analysis-options title="Permanent link">&para;</a></h4> <p>Each of the following arguments are optional, and do not need to be provided.</p> <p><code>--aggregate {mapped, none}</code></p> <blockquote> <p><strong>Cell Ranger aggregate normalization.</strong> <br> <em>type: string</em></p> <p>This option defines the normalization mode that should be used. Mapped is what Cell Ranger would run by default, which subsamples reads from higher depth samples until each library type has an equal number of reads per cell that are confidently mapped. None means to not normalize at all. If this flag is not used then aggregate will not be run. Aggregate analysis is generally not needed, but it can be used to generate a Loupe Browser file for interactive exploration of the data. To run Cell Ranger aggregate, please select one of the following options: mapped, none.</p> <p><strong><em>Example:</em></strong> <code>--aggregate mapped</code></p> </blockquote> <hr> <p><code>--exclude-introns</code></p> <blockquote> <p><strong>Exclude introns from the count alignment.</strong> <br> <em>type: boolean flag</em></p> <p>Turns off the option of including introns when performing alignment. This flag is only applicable when dealing with gene expression related data.</p> <p><strong><em>Example:</em></strong> <code>--exclude-introns</code></p> </blockquote> <hr> <p><code>--create-bam</code></p> <blockquote> <p><strong>Create bam files.</strong> <br> <em>type: boolean flag</em></p> <p>By default the no-bam flag is used when running Cell Ranger. Use this flag to ensure that a bam file is created for each sample during analysis. This flag is only applicable when dealing with gene expression related data.</p> <p><strong><em>Example:</em></strong> <code>--create-bam</code></p> </blockquote> <hr> <p><code>--filter FILTER</code></p> <blockquote> <p><strong>Filter threshold file.</strong> <br> <em>type: file</em></p> <p>Filter threshold file. A CSV file containing the different thresholds to be applied for individual samples within the project during the QC analysis. The file should contain a header row with Sample as the column name for the sample IDs, and the name of each metric that will be filtered along with if it is the high or low threshold for that metric. Each row is then the entries for each sample that the manual thresholds will be applied. If no file is provided then the default thresholds will be used. If a cell is left blank for a sample then that sample would not be filtered based on that criteria. This flag is currently only applicable when dealing with GEX projects.</p> <p><em>Here is an example filter.csv file:</em> <div class=highlight><pre><span></span><code>Sample,nFeature_RNA_low,nFeature_RNA_high,percent.mito_high
sample1,500,6000,15
sample2,500,6000,5
sample4,500,6000,5
</code></pre></div></p> <p><em>Where:</em> </p> <ul> <li><em>Sample:</em> Unique sample ID that should match the sample name used for Cell Ranger count.</li> <li><em>nFeature_RNA_low,nFeature_RNA_high,percent.mito_high:</em> Example entries that can be used for manual thresholding. The column names need to be formatted as metadataname_high/low. Entries that ends with high will be treated as the upper threshold. Entries that ends with low will be treated as the lower threshold. Valid metadata names include nCount_RNA, nFeature_RNA, and percent.mito.</li> </ul> <p><strong><em>Example:</em></strong> <code>--filter filter.csv</code></p> </blockquote> <hr> <p><code>--metadata METADATA</code></p> <blockquote> <p><strong>Sample metadata file.</strong> <br> <em>type: file</em></p> <p>Sample metadata file. A CSV file containing sample level metadata information that will be included as new metadata columns during QC analysis. The file should contain a header row with Sample as the column name for the sample IDs, and the name of each metadata column that will be added to their associated samples. Each row is then the entries for each sample with the values that will be included as metadata. If no file is provided then no metadata will be added to the samples. If a cell is left blank for a sample then the metadata column for that sample would be an empty string. This flag is currently only applicable when dealing with GEX projects.</p> <p><em>Here is an example metadata.csv file:</em> <div class=highlight><pre><span></span><code>Sample,Type,batch
sample1,tumor,1
sample2,normal,1
sample4,tumor,2
</code></pre></div></p> <p><em>Where:</em> </p> <ul> <li><em>Sample:</em> Unique sample ID that should match the sample name used for Cell Ranger count.</li> <li><em>Type,batch:</em> Example metadata entries that can be applied to the created Seurat object. The column names will be the resulting metadata entries, so it is recommended to use ones that do not overlap with column names that would be created automatically.</li> </ul> <p><strong><em>Example:</em></strong> <code>--metadata metadata.csv</code></p> </blockquote> <hr> <p><code>--rename RENAME</code></p> <blockquote> <p><strong>Rename sample file.</strong> <br> <em>type: file</em></p> <p>Rename sample file. A CSV file containing the name of the FASTQ file and the new name of the sample. Only the samples listed in the CSV files will be run.</p> <p><em>Here is an example rename.csv file:</em> <div class=highlight><pre><span></span><code>FASTQ,Name
original_name1,new_name1
original_name2,new_name2
original_name3,new_name3
original_name3-2,new_name3
original_name4,original_name4
</code></pre></div></p> <p><em>Where:</em></p> <ul> <li><em>FASTQ:</em> The name that is used in the FASTQ file</li> <li><em>Name:</em> Unique sample ID that is the sample name used for Cell Ranger count.</li> </ul> <p>In this example, new_name3 has FASTQ files with two different names. With this input, both sets of FASTQ files will be used when processing the sample as new_name3. original_name4 will not be renamed. Any FASTQ file that does not have the name original_name1, original_name2, original_name3, or original_name4 will not be run.</p> <p><strong><em>Example:</em></strong> <code>--rename rename.csv</code></p> </blockquote> <hr> <p><code>--forcecells FORCECELLS</code></p> <blockquote> <p><strong>Force cells file.</strong><br> <em>type: file</em></p> <p>Force cells file. A CSV file containing the name of the sample (the Cell Ranger outputted name) and the number of cells to force the sample to. It will generally be used if the first analysis run appears to do a poor job at estimating the number of cells, and a re-run is needed to adjust the number of cells in the sample.</p> <p><em>Here is an example forcecells.csv file:</em> <div class=highlight><pre><span></span><code>Sample,Cells
Sample1,3000
Sample2,5000
</code></pre></div></p> <p><em>Where:</em></p> <ul> <li><em>Sample:</em> The sample name used as the Cell Ranger output</li> <li><em>Cells:</em> The number of cells the sample should be forced to</li> </ul> <p>In this example, Sample1 and Sample2 will be run while being forced to have 3000 and 5000 cells respectively. Any other samples that are processed will be run without using the force cells flag and will use the default cell calling algorithm.</p> <p><strong><em>Example:</em></strong> <code>--forcecells forcecells.csv</code></p> </blockquote> <h3 id=22-vdj>2.2 VDJ<a class=headerlink href=#22-vdj title="Permanent link">&para;</a></h3> <h4 id=221-required-arguments>2.2.1 Required Arguments<a class=headerlink href=#221-required-arguments title="Permanent link">&para;</a></h4> <p>Each of the following arguments are required. Failure to provide a required argument will result in a non-zero exit-code.</p> <p><code>--input INPUT [INPUT ...]</code> </p> <blockquote> <p><strong>Input FastQ file(s) or Cell Ranger folder(s).</strong><br> <em>type: file(s) or folder(s)</em> </p> <p>FastQ Input: One or more FastQ files can be provided. The pipeline does NOT support single-end data. From the command-line, each input file should separated by a space. Multiple input FastQ files per sample can be provided. Globbing is supported! This makes selecting FastQ files easy. Input FastQ files should always be gzipp-ed.</p> <p><strong><em>Example:</em></strong> <code>--input .tests/*.R?.fastq.gz</code></p> <p>Cell Ranger Input: Cell Ranger output folders can be provided. It is expected that the outs folder is contained within the Cell Ranger output folders, and keep the normal output folder structure. Globbing is supported!</p> <p><strong><em>Example:</em></strong> <code>--input .tests/*/</code></p> </blockquote> <hr> <p><code>--output OUTPUT</code></p> <blockquote> <p><strong>Path to an output directory.</strong> <br> <em>type: path</em></p> <p>This location is where the pipeline will create all of its output files, also known as the pipeline's working directory. If the provided output directory does not exist, it will be created automatically.</p> <p><strong><em>Example:</em></strong> <code>--output /data/$USER/cell-seek_out</code></p> </blockquote> <hr> <p><code>--pipeline vdj</code></p> <blockquote> <p><strong>The pipeline to run.</strong> <br> <em>type: string</em></p> <p>This option selects the version of the pipeline to run. The documentation provided is based on choosing the option for VDJ.</p> <p><strong><em>Example:</em></strong> <code>--pipeline vdj</code></p> </blockquote> <hr> <p><code>--genome {hg38, mm10, custom.json}</code></p> <blockquote> <p><strong>Reference genome.</strong> <br> <em>type: string</em></p> <p>This option defines the reference genome of the samples. cell-seek does comes bundled with prebuilt reference files for human and mouse samples, e.g. hg38 or mm10. Since there is no 2024 release VDJ reference, if hg2024 or mm2024 is selected the VDJ reference CR 7.1 release will be used for human, and CR 7.0 release will be used for mouse.</p> <p>A custom reference genome can also be provided via a json file. Additional information for creating this json file can be found in <a href=../genome/ ><code>cell-seek <b>genome</b></code></a>.</p> <p>For prebuilt references please select one of the following options: hg38, mm10</p> <p><strong><em>Example:</em></strong> <code>--genome hg38</code></p> </blockquote> <hr> <p><code>--cellranger {7.1.0, 7.2.0, 8.0.0, 9.0.0, 10.0.0}</code></p> <blockquote> <p><strong>The version of Cell Ranger to run.</strong> <br> <em>type: string</em></p> <p>This option specifies which version of Cell Ranger to use when running GEX, VDJ, CITE, or MULTI pipelines. Please select one of the following options: 7.1.0, 7.2.0, 8.0.0, 9.0.0, 10.0.0</p> <p><strong><em>Example:</em></strong> <code>--cellranger 7.1.0</code></p> </blockquote> <h4 id=222-analysis-options>2.2.2 Analysis Options<a class=headerlink href=#222-analysis-options title="Permanent link">&para;</a></h4> <p><code>--rename RENAME</code></p> <blockquote> <p><strong>Rename sample file.</strong><br> <em>type: file</em></p> <p>Rename sample file. A CSV file containing the name of the FASTQ file and the new name of the sample. Only the samples listed in the CSV files will be run.</p> <p><em>Here is an example rename.csv file:</em> <div class=highlight><pre><span></span><code>FASTQ,Name
original_name1,new_name1
original_name2,new_name2
original_name3,new_name3
original_name3-2,new_name3
original_name4,original_name4
</code></pre></div></p> <p><em>Where:</em></p> <ul> <li><em>FASTQ:</em> The name that is used in the FASTQ file</li> <li><em>Name:</em> Unique sample ID that is the sample name used for Cell Ranger count.</li> </ul> <p>In this example, new_name3 has FASTQ files with two different names. With this input, both sets of FASTQ files will be used when processing the sample as new_name3. original_name4 will not be renamed. Any FASTQ file that does not have the name original_name1, original_name2, original_name3, or original_name4 will not be run.</p> <p><strong><em>Example:</em></strong> <code>--rename rename.csv</code></p> </blockquote> <h3 id=23-cite>2.3 CITE<a class=headerlink href=#23-cite title="Permanent link">&para;</a></h3> <h4 id=231-required-arguments>2.3.1 Required Arguments<a class=headerlink href=#231-required-arguments title="Permanent link">&para;</a></h4> <p>Each of the following arguments are required. Failure to provide a required argument will result in a non-zero exit-code.</p> <p><code>--input INPUT [INPUT ...]</code> </p> <blockquote> <p><strong>Input FastQ file(s) or Cell Ranger folder(s).</strong><br> <em>type: file(s) or folder(s)</em> </p> <p>FastQ Input: One or more FastQ files can be provided. The pipeline does NOT support single-end data. From the command-line, each input file should separated by a space. Multiple input FastQ files per sample can be provided. Globbing is supported! This makes selecting FastQ files easy. Input FastQ files should always be gzipp-ed.</p> <p><strong><em>Example:</em></strong> <code>--input .tests/*.R?.fastq.gz</code></p> <p>Cell Ranger Input: Cell Ranger output folders can be provided. It is expected that the outs folder is contained within the Cell Ranger output folders, and keep the normal output folder structure. Globbing is supported!</p> <p><strong><em>Example:</em></strong> <code>--input .tests/*/</code></p> </blockquote> <hr> <p><code>--output OUTPUT</code></p> <blockquote> <p><strong>Path to an output directory.</strong> <br> <em>type: path</em></p> <p>This location is where the pipeline will create all of its output files, also known as the pipeline's working directory. If the provided output directory does not exist, it will be created automatically.</p> <p><strong><em>Example:</em></strong> <code>--output /data/$USER/cell-seek_out</code></p> </blockquote> <hr> <p><code>--pipeline cite</code></p> <blockquote> <p><strong>The pipeline to run.</strong> <br> <em>type: string</em></p> <p>This option selects the version of the pipeline to run. The documentation provided is based on choosing the option for CITE-seek.</p> <p><strong><em>Example:</em></strong> <code>--pipeline cite</code></p> </blockquote> <hr> <p><code>--genome {hg38, mm10, hg2024, mm2024, custom.json}</code></p> <blockquote> <p><strong>Reference genome.</strong> <br> <em>type: string</em></p> <p>This option defines the reference genome of the samples. cell-seek does comes bundled with prebuilt reference files for human and mouse samples, The options hg38 or mm10 would select the 2020 release of the reference. The options hg2024 or mm2024 would select the 2024 release of the reference. More information about the officially released references can be found on the <a href=https://www.10xgenomics.com/support/software/cell-ranger/latest/release-notes/cr-reference-release-notes>10x Genomics website</a>.</p> <p>A custom reference genome can also be provided via a json file. Additional information for creating this json file can be found in <a href=../genome/ ><code>cell-seek <b>genome</b></code></a>.</p> <p>For prebuilt references please select one of the following options: hg38, mm10, hg2024, mm2024</p> <p><strong><em>Example:</em></strong> <code>--genome hg38</code></p> </blockquote> <hr> <p><code>--cellranger {7.1.0, 7.2.0, 8.0.0, 9.0.0, 10.0.0}</code></p> <blockquote> <p><strong>The version of Cell Ranger to run.</strong> <br> <em>type: string</em></p> <p>This option specifies which version of Cell Ranger to use when running GEX, VDJ, CITE, or MULTI pipelines. Please select one of the following options: 7.1.0, 7.2.0, 8.0.0, 9.0.0, 10.0.0</p> <p><strong><em>Example:</em></strong> <code>--cellranger 7.1.0</code></p> </blockquote> <h4 id=232-conditionally-required-arguments>2.3.2 Conditionally Required Arguments<a class=headerlink href=#232-conditionally-required-arguments title="Permanent link">&para;</a></h4> <p>The following arguments are only required when FastQ files are used as input. They are not required when Cell Ranger output file is used as input.</p> <p><code>--libraries LIBRARIES</code></p> <blockquote> <p><strong>Libraries file.</strong> <br> <em>type: file</em></p> <p>A CSV file containing information about each library. It contains each sample's name, flowcell, demultiplexed name, and library type. More information about the libraries file and its requirements can be found on the <a href=https://www.10xgenomics.com/support/software/cell-ranger/analysis/running-pipelines/cr-feature-bc-analysis>10x Genomics website</a>.</p> <p><em>Here is an example libraries.csv file:</em> <div class=highlight><pre><span></span><code>Name,Flowcell,Sample,Type
IL15_LNs,H7CNNBGXG,IL15_LNs,Gene Expression
IL15_LNs,H7CT7BGXG,IL15_LNs_ADT,Antibody Capture
EGR1,H7CNNBGXG,EGR1_GEX,Gene Expression
EGR1,H7CT7BGXG,EGR1_ADT,Antibody Capture
</code></pre></div></p> <p><em>Where:</em></p> <ul> <li><em>Name:</em> name of the sample passed to Cell Ranger. </li> <li><em>Flowcell:</em> The flowcell ID that contains the FASTQ files for this set of data. </li> <li><em>Sample:</em> Name that was used when demultiplexing, this should match the FASTQ files. </li> <li><em>Type:</em> library type for each sample. List of supported options: <ul> <li>Gene Expression</li> <li>CRISPR Guide Capture</li> <li>Antibody Capture</li> <li>Custom</li> </ul> </li> </ul> <p><strong><em>Example:</em></strong> <code>--libraries libraries.csv</code></p> </blockquote> <hr> <p><code>--features FEATURES</code></p> <blockquote> <p><strong>Features file.</strong> <br> <em>type: file</em></p> <p>A feature reference CSV file containing information for processing a feature barcode data. This file should contain a unique ID for the feature, a human readable name, sequence, feature type, read, and pattern. More information about the libraries file and its requirements can be found on the <a href=https://www.10xgenomics.com/support/software/cell-ranger/analysis/running-pipelines/cr-feature-bc-analysis>10x Genomics website</a>.</p> <p><em>Here is an example features.csv file:</em> <div class=highlight><pre><span></span><code>id,name,sequence,feature_type,read,pattern
CITE_CD64,CD64,AGTGGG,Antibody Capture,R2,5PNN(BC)
CITE_CD8,CD8,TCACCGT,Antibody Capture,R2,5PNNN(BC)
</code></pre></div></p> <p><em>Where:</em> </p> <ul> <li><em>id:</em> Unique ID for this feature. Must not contain whitespace, quote or comma characters. Each ID must be unique and must not collide with a gene identifier from the transcriptome.</li> <li><em>name:</em> Human-readable name for this feature. Must not contain whitespace.</li> <li><em>sequence:</em> Nucleotide barcode sequence associated with this feature, e.g. the antibody barcode or sgRNA protospacer sequence.</li> <li><em>read:</em> Specifies which RNA sequencing read contains the Feature Barcode sequence. Must be R1 or R2, but in most cases R2 is the correct read.</li> <li> <p><em>pattern:</em> Specifies how to extract the sequence of the feature barcode from the read.</p> </li> <li> <p><em>Type:</em> Type of the feature. List of supported options: </p> <ul> <li>Antibody Capture</li> <li>CRISPR Guide Capture</li> <li>Antigen Capture</li> <li>Custom</li> </ul> </li> </ul> <p><strong><em>Example:</em></strong> <code>--features features.csv</code></p> </blockquote> <h4 id=233-analysis-options>2.3.3 Analysis Options<a class=headerlink href=#233-analysis-options title="Permanent link">&para;</a></h4> <p><code>--exclude-introns</code></p> <blockquote> <p><strong>Exclude introns from the count alignment.</strong> <br> <em>type: boolean flag</em></p> <p>Turns off the option of including introns when performing alignment. This flag is only applicable when dealing with gene expression related data.</p> <p><strong><em>Example:</em></strong> <code>--exclude-introns</code></p> </blockquote> <hr> <p><code>--create-bam</code></p> <blockquote> <p><strong>Create bam files.</strong> <br> <em>type: boolean flag</em></p> <p>By default the no-bam flag is used when running Cell Ranger. Use this flag to ensure that a bam file is created for each sample during analysis. This flag is only applicable when dealing with gene expression related data.</p> <p><strong><em>Example:</em></strong> <code>--create-bam</code></p> </blockquote> <hr> <p><code>--forcecells FORCECELLS</code></p> <blockquote> <p><strong>Force cells file.</strong><br> <em>type: file</em></p> <p>Force cells file. A CSV file containing the name of the sample (the Cell Ranger outputted name) and the number of cells to force the sample to. It will generally be used if the first analysis run appears to do a poor job at estimating the number of cells, and a re-run is needed to adjust the number of cells in the sample.</p> <p><em>Here is an example forcecells.csv file:</em> <div class=highlight><pre><span></span><code>Sample,Cells
Sample1,3000
Sample2,5000
</code></pre></div></p> <p><em>Where:</em></p> <ul> <li><em>Sample:</em> The sample name used as the Cell Ranger output</li> <li><em>Cells:</em> The number of cells the sample should be forced to</li> </ul> <p>In this example, Sample1 and Sample2 will be run while being forced to have 3000 and 5000 cells respectively. Any other samples that are processed will be run without using the force cells flag and will use the default cell calling algorithm.</p> <p><strong><em>Example:</em></strong> <code>--forcecells forcecells.csv</code></p> </blockquote> <h3 id=24-multi>2.4 MULTI<a class=headerlink href=#24-multi title="Permanent link">&para;</a></h3> <p>There are multiple different combinations of library types that may result in the use of Cell Ranger <code>multi</code> analysis. Any combination that combines GEX and VDJ data for cell calls, or the use of HTO with the Cell Ranger hashtag caller would need <code>multi</code> analysis.</p> <h4 id=241-required-arguments>2.4.1 Required Arguments<a class=headerlink href=#241-required-arguments title="Permanent link">&para;</a></h4> <p>Each of the following arguments are required. Failure to provide a required argument will result in a non-zero exit-code.</p> <p><code>--input INPUT [INPUT ...]</code> </p> <blockquote> <p><strong>Input FastQ file(s) or Cell Ranger folder(s).</strong><br> <em>type: file(s) or folder(s)</em> </p> <p>FastQ Input: One or more FastQ files can be provided. The pipeline does NOT support single-end data. From the command-line, each input file should separated by a space. Multiple input FastQ files per sample can be provided. Globbing is supported! This makes selecting FastQ files easy. Input FastQ files should always be gzipp-ed.</p> <p><strong><em>Example:</em></strong> <code>--input .tests/*.R?.fastq.gz</code></p> <p>Cell Ranger Input: Cell Ranger output folders can be provided. It is expected that the outs folder is contained within the Cell Ranger output folders, and keep the normal output folder structure. Globbing is supported!</p> <p><strong><em>Example:</em></strong> <code>--input .tests/*/</code></p> </blockquote> <hr> <p><code>--output OUTPUT</code></p> <blockquote> <p><strong>Path to an output directory.</strong> <br> <em>type: path</em></p> <p>This location is where the pipeline will create all of its output files, also known as the pipeline's working directory. If the provided output directory does not exist, it will be created automatically.</p> <p><strong><em>Example:</em></strong> <code>--output /data/$USER/cell-seek_out</code></p> </blockquote> <hr> <p><code>--pipeline multi</code></p> <blockquote> <p><strong>The pipeline to run.</strong> <br> <em>type: string</em></p> <p>This option selects the version of the pipeline to run. The documentation provided is based on choosing the option for multi analysis.</p> <p><strong><em>Example:</em></strong> <code>--pipeline multi</code></p> </blockquote> <hr> <p><code>--genome {hg38, mm10, hg2024, mm2024, custom.json}</code></p> <blockquote> <p><strong>Reference genome.</strong> <br> <em>type: string</em></p> <p>This option defines the reference genome of the samples. cell-seek does comes bundled with prebuilt reference files for human and mouse samples, The options hg38 or mm10 would select the 2020 release of the reference. The options hg2024 or mm2024 would select the 2024 release of the reference. More information about the officially released references can be found on the <a href=https://www.10xgenomics.com/support/software/cell-ranger/latest/release-notes/cr-reference-release-notes>10x Genomics website</a>. Since there is no 2024 released VDJ reference, if hg2024 or mm2024 is selected in a run that includes VDJ data, the VDJ reference CR 7.1 release will be used for human, and CR 7.0 release will be used for mouse.</p> <p>A custom reference genome can also be provided via a json file. Additional information for creating this json file can be found in <a href=../genome/ ><code>cell-seek <b>genome</b></code></a>.</p> <p>For prebuilt references please select one of the following options: hg38, mm10, hg2024, mm2024</p> <p><strong><em>Example:</em></strong> <code>--genome hg38</code></p> </blockquote> <hr> <p><code>--cellranger {7.1.0, 7.2.0, 8.0.0, 9.0.0, 10.0.0}</code></p> <blockquote> <p><strong>The version of Cell Ranger to run.</strong> <br> <em>type: string</em></p> <p>This option specifies which version of Cell Ranger to use when running GEX, VDJ, CITE, or MULTI pipelines. Please select one of the following options: 7.1.0, 7.2.0, 8.0.0, 9.0.0, 10.0.0</p> <p><strong><em>Example:</em></strong> <code>--cellranger 7.1.0</code></p> </blockquote> <h4 id=242-conditionally-required-arguments>2.4.2 Conditionally Required Arguments<a class=headerlink href=#242-conditionally-required-arguments title="Permanent link">&para;</a></h4> <p>The following arguments are only required when FastQ files are used as input. They are not required when Cell Ranger output file is used as input.</p> <p><code>--libraries LIBRARIES</code></p> <blockquote> <p><strong>Libraries file.</strong> <br> <em>type: file</em></p> <p>A CSV file containing information about each library. It contains each sample's name, flowcell, demultiplexed name, and library type. More information about the libraries file and its requirements can be found on the <a href=https://www.10xgenomics.com/support/software/cell-ranger/analysis/running-pipelines/cr-feature-bc-analysis>10x Genomics website</a>.</p> <p><em>Here is an example libraries.csv file:</em> <div class=highlight><pre><span></span><code>Name,Flowcell,Sample,Type
IL15_LNs,H7CNNBGXG,IL15_LNs,Gene Expression
IL15_LNs,H7CT7BGXG,IL15_LNs_BC,Antibody Capture
</code></pre></div></p> <p><em>Where:</em></p> <ul> <li><em>Name:</em> name of the sample passed to Cell Ranger. </li> <li><em>Flowcell:</em> The flowcell ID that contains the FASTQ files for this set of data. </li> <li><em>Sample:</em> Name that was used when demultiplexing, this should match the FASTQ files. </li> <li><em>Type:</em> library type for each sample. List of supported options: <ul> <li>Gene Expression</li> <li>CRISPR Guide Capture</li> <li>Antibody Capture</li> <li>Multiplexing Capture</li> <li>VDJ</li> <li>Custom</li> </ul> </li> </ul> <p><strong><em>Example:</em></strong> <code>--libraries libraries.csv</code></p> </blockquote> <h4 id=243-analysis-options>2.4.3 Analysis Options<a class=headerlink href=#243-analysis-options title="Permanent link">&para;</a></h4> <p>Each of the following arguments are optional, and do not need to be provided.</p> <p><code>--features FEATURES</code></p> <blockquote> <p><strong>Features file.</strong> <br> <em>type: file</em></p> <p>A feature reference CSV file containing information for processing a feature barcode data. This file is used in feature barcode, and may be used in multi analysis. This file should contain a unique ID for the feature, a human readable name, sequence, feature type, read, and pattern. More information about the libraries file and its requirements can be found on the <a href=https://www.10xgenomics.com/support/software/cell-ranger/analysis/running-pipelines/cr-feature-bc-analysis>10x Genomics website</a>.</p> <p><em>Here is an example features.csv file:</em> <div class=highlight><pre><span></span><code>id,name,sequence,feature_type,read,pattern
CITE_CD64,CD64,AGTGGG,Antibody Capture,R2,5PNN(BC)
CITE_CD8,CD8,TCACCGT,Antibody Capture,R2,5PNNN(BC)
</code></pre></div></p> <p><em>Where:</em> </p> <ul> <li><em>id:</em> Unique ID for this feature. Must not contain whitespace, quote or comma characters. Each ID must be unique and must not collide with a gene identifier from the transcriptome.</li> <li><em>name:</em> Human-readable name for this feature. Must not contain whitespace.</li> <li><em>sequence:</em> Nucleotide barcode sequence associated with this feature, e.g. the antibody barcode or sgRNA protospacer sequence.</li> <li><em>read:</em> Specifies which RNA sequencing read contains the Feature Barcode sequence. Must be R1 or R2, but in most cases R2 is the correct read.</li> <li> <p><em>pattern:</em> Specifies how to extract the sequence of the feature barcode from the read.</p> </li> <li> <p><em>Type:</em> Type of the feature. List of supported options: </p> <ul> <li>CRISPR Guide Capture</li> <li>Antibody Capture</li> <li>Custom</li> </ul> </li> </ul> <p><strong><em>Example:</em></strong> <code>--features features.csv</code></p> </blockquote> <hr> <p><code>--cmo-reference CMOREFERENCE</code></p> <blockquote> <p><strong>CMO reference file.</strong> <br> <em>type: file</em></p> <p>A CMO reference CSV file containing information for processing hashtag data, which is used in multi analysis if custom hashtags will be processed. This file should contain a unique ID for the hashtag, a human readable name, sequence, feature type, read, and pattern. More information about the cmo reference file and its requirements can be found on the <a href=https://www.10xgenomics.com/support/software/cell-ranger/analysis/cr-3p-multi>10x Genomics website</a>.</p> <p><em>Here is an example cmo_reference.csv file:</em> <div class=highlight><pre><span></span><code>id,name,sequence,feature_type,read,pattern
CMO301,CMO301,ATGAGGAATTCCTGC,Multiplexing Capture,R2,5P(BC)
CMO302,CMO302,CATGCCAATAGAGCG,Multiplexing Capture,R2,5P(BC)
</code></pre></div></p> <p><em>Where:</em> </p> <ul> <li><em>id:</em> Unique ID for this feature. Must not contain whitespace, quote or comma characters. Each ID must be unique and must not collide with a gene identifier from the transcriptome.</li> <li><em>name:</em> Human-readable name for this feature. Must not contain whitespace.</li> <li><em>sequence:</em> Nucleotide barcode sequence associated with this hashtag</li> <li><em>feature_type:</em> Type of the feature. This should always be multiplexing capture.</li> <li><em>read:</em> Specifies which RNA sequencing read contains the Feature Barcode sequence. Must be R1 or R2, but in most cases R2 is the correct read.</li> <li><em>pattern:</em> Specifies how to extract the sequence of the feature barcode from the read.</li> </ul> <p><strong><em>Example:</em></strong> <code>--cmo-reference cmo_reference.csv</code></p> </blockquote> <hr> <p><code>--cmo-sample CMOSAMPLE</code></p> <blockquote> <p><strong>CMO sample file.</strong> <br> <em>type: file</em></p> <p>A CMO sample CSV file containing sample to hashtag information used for multi analysis. CMO IDs should match the ones used in the CMO reference file. If no CMO reference is provided then CMO ID should match the ones used by 10x. The same CMO sample will be used on all multi libraries. This file should contain a unique sample ID for the sample, and the CMO IDs associated with that sample. If more than one CMO ID is associated with a sample then a | should be used to separate the tags. More information and examples about the samples section of the multi config file and its requirements can be found on the <a href=https://www.10xgenomics.com/support/software/cell-ranger/analysis/cr-3p-multi>10x Genomics website</a>.</p> <p><em>Here is an example cmo_sample.csv file:</em> <div class=highlight><pre><span></span><code>sample_id,cmo_ids
sample1,CMO301
sample2,CMO302|CMO303
</code></pre></div></p> <p><em>Where:</em> </p> <ul> <li><em>sample_id:</em> Unique sample ID for this hashtagged sample. Must not contain whitespace, quote or comma characters. Each sample ID must be unique.</li> <li><em>cmo_ids:</em> Unique CMO ID(s) that the sample is hashtagged with. Must match either entries in the cmo_reference.csv file or 10x CMO IDs.</li> </ul> <p><strong><em>Example:</em></strong> <code>--cmo-sample cmo_sample.csv</code></p> </blockquote> <hr> <p><code>--exclude-introns</code></p> <blockquote> <p><strong>Exclude introns from the count alignment.</strong> <br> <em>type: boolean flag</em></p> <p>Turns off the option of including introns when performing alignment. This flag is only applicable when dealing with gene expression related data.</p> <p><strong><em>Example:</em></strong> <code>--exclude-introns</code></p> </blockquote> <hr> <p><code>--create-bam</code></p> <blockquote> <p><strong>Create bam files.</strong> <br> <em>type: boolean flag</em></p> <p>By default the no-bam flag is used when running Cell Ranger. Use this flag to ensure that a bam file is created for each sample during analysis. This flag is only applicable when dealing with gene expression related data.</p> <p><strong><em>Example:</em></strong> <code>--create-bam</code></p> </blockquote> <hr> <p><code>--forcecells FORCECELLS</code></p> <blockquote> <p><strong>Force cells file.</strong><br> <em>type: file</em></p> <p>Force cells file. A CSV file containing the name of the sample (the Cell Ranger outputted name) and the number of cells to force the sample to. It will generally be used if the first analysis run appears to do a poor job at estimating the number of cells, and a re-run is needed to adjust the number of cells in the sample.</p> <p>This file can created in two different formats. The first one will contain the name of the sample and the number of cells to be forced to.</p> <p><em>Here is an example forcecells.csv file:</em> <div class=highlight><pre><span></span><code>Sample,Cells
Sample1,3000
Sample2,5000
</code></pre></div></p> <p><em>Where:</em></p> <ul> <li><em>Sample:</em> The sample name used as the Cell Ranger output</li> <li><em>Cells:</em> The number of cells the sample should be forced to</li> </ul> <p>In this example, Sample1 and Sample2 will be run while being forced to have 3000 and 5000 cells respectively. Any other samples that are processed will be run without using the force cells flag and will use the default cell calling algorithm.</p> <p>The second format is only compatible when hashtag multiplexing is used and the number of cells needs to be forced for a specific hashtagged sample.</p> <p><em>Here is an example forcecells.csv file:</em> <div class=highlight><pre><span></span><code>Name,Sample,Cells
Library1,Sample1,3000
Library1,Sample2,5000
</code></pre></div></p> <p><em>Where:</em></p> <ul> <li><em>Library:</em> The name of the library that is provided as to Cell Ranger when running multi analysis. This should match the name that is given in the libraries.csv file.</li> <li><em>Sample:</em> The sample ID used for the associated hashtag. This will have to match the value used in the CMO sample file or the CMO reference file that is provided as input. If only a CMO reference file is provided, the pipeline default assigns each hashtag with the IDs of HTO_1, HTO_2, etc.</li> <li><em>Cells:</em> The number of cells the sample should be forced to</li> </ul> <p>In this example, the hashtags HTO_1 and HTO_2 in Library 1 will be run while being forced to 3000 and 5000 cells respectively. Any other libraries or samples that are processed will be run without using the force cells flag.</p> <p><strong><em>Example:</em></strong> <code>--forcecells forcecells.csv</code></p> </blockquote> <h3 id=25-atac>2.5 ATAC<a class=headerlink href=#25-atac title="Permanent link">&para;</a></h3> <h4 id=251-required-arguments>2.5.1 Required Arguments<a class=headerlink href=#251-required-arguments title="Permanent link">&para;</a></h4> <p>Each of the following arguments are required. Failure to provide a required argument will result in a non-zero exit-code.</p> <p><code>--input INPUT [INPUT ...]</code> </p> <blockquote> <p><strong>Input FastQ file(s) or Cell Ranger folder(s).</strong><br> <em>type: file(s) or folder(s)</em> </p> <p>FastQ Input: One or more FastQ files can be provided. The pipeline does NOT support single-end data. From the command-line, each input file should separated by a space. Multiple input FastQ files per sample can be provided. Globbing is supported! This makes selecting FastQ files easy. Input FastQ files should always be gzipp-ed.</p> <p><strong><em>Example:</em></strong> <code>--input .tests/*.R?.fastq.gz</code></p> <p>Cell Ranger Input: Cell Ranger output folders can be provided. It is expected that the outs folder is contained within the Cell Ranger output folders, and keep the normal output folder structure. Globbing is supported!</p> <p><strong><em>Example:</em></strong> <code>--input .tests/*/</code></p> </blockquote> <hr> <p><code>--output OUTPUT</code></p> <blockquote> <p><strong>Path to an output directory.</strong> <br> <em>type: path</em></p> <p>This location is where the pipeline will create all of its output files, also known as the pipeline's working directory. If the provided output directory does not exist, it will be created automatically.</p> <p><strong><em>Example:</em></strong> <code>--output /data/$USER/cell-seek_out</code></p> </blockquote> <hr> <p><code>--pipeline atac</code></p> <blockquote> <p><strong>The pipeline to run.</strong> <br> <em>type: string</em></p> <p>This option selects the version of the pipeline to run. The documentation provided is based on choosing the option for ATAC.</p> <p><strong><em>Example:</em></strong> <code>--pipeline atac</code></p> </blockquote> <hr> <p><code>--genome {hg38, mm10, hg2024, mm2024, custom.json}</code></p> <blockquote> <p><strong>Reference genome.</strong> <br> <em>type: string</em></p> <p>This option defines the reference genome of the samples. cell-seek does comes bundled with prebuilt reference files for human and mouse samples, e.g. hg38 or mm10.</p> <p>A custom reference genome can also be provided via a json file. Additional information for creating this json file can be found in <a href=../genome/ ><code>cell-seek <b>genome</b></code></a>.</p> <p>For prebuilt references please select one of the following options: hg38, mm10</p> <p><strong><em>Example:</em></strong> <code>--genome hg38</code></p> </blockquote> <h4 id=252-analysis-options>2.5.2 Analysis Options<a class=headerlink href=#252-analysis-options title="Permanent link">&para;</a></h4> <p><code>--rename RENAME</code></p> <blockquote> <p><strong>Rename sample file.</strong> <em>type: file</em></p> <p>Rename sample file. A CSV file containing the name of the FASTQ file and the new name of the sample. Only the samples listed in the CSV files will be run.</p> <p><em>Here is an example rename.csv file:</em> <div class=highlight><pre><span></span><code>FASTQ,Name
original_name1,new_name1
original_name2,new_name2
original_name3,new_name3
original_name3-2,new_name3
original_name4,original_name4
</code></pre></div></p> <p><em>Where:</em></p> <ul> <li><em>FASTQ:</em> The name that is used in the FASTQ file</li> <li><em>Name:</em> Unique sample ID that is the sample name used for Cell Ranger count.</li> </ul> <p>In this example, new_name3 has FASTQ files with two different names. With this input, both sets of FASTQ files will be used when processing the sample as new_name3. original_name4 will not be renamed. Any FASTQ file that does not have the name original_name1, original_name2, original_name3, or original_name4 will not be run.</p> <p><strong><em>Example:</em></strong> <code>--rename rename.csv</code></p> </blockquote> <hr> <p><code>--forcecells FORCECELLS</code></p> <blockquote> <p><strong>Force cells file.</strong><br> <em>type: file</em></p> <p>Force cells file. A CSV file containing the name of the sample (the Cell Ranger outputted name) and the number of cells to force the sample to. It will generally be used if the first analysis run appears to do a poor job at estimating the number of cells, and a re-run is needed to adjust the number of cells in the sample.</p> <p><em>Here is an example forcecells.csv file:</em> <div class=highlight><pre><span></span><code>Sample,Cells
Sample1,3000
Sample2,5000
</code></pre></div></p> <p><em>Where:</em></p> <ul> <li><em>Sample:</em> The sample name used as the Cell Ranger output</li> <li><em>Cells:</em> The number of cells the sample should be forced to</li> </ul> <p>In this example, Sample1 and Sample2 will be run while being forced to have 3000 and 5000 cells respectively. Any other samples that are processed will be run without using the force cells flag and will use the default cell calling algorithm.</p> <p><strong><em>Example:</em></strong> <code>--forcecells forcecells.csv</code></p> </blockquote> <h3 id=26-multiome>2.6 Multiome<a class=headerlink href=#26-multiome title="Permanent link">&para;</a></h3> <h4 id=261-required-arguments>2.6.1 Required Arguments<a class=headerlink href=#261-required-arguments title="Permanent link">&para;</a></h4> <p>Each of the following arguments are required. Failure to provide a required argument will result in a non-zero exit-code.</p> <p><code>--input INPUT [INPUT ...]</code> </p> <blockquote> <p><strong>Input FastQ file(s) or Cell Ranger folder(s).</strong><br> <em>type: file(s) or folder(s)</em> </p> <p>FastQ Input: One or more FastQ files can be provided. The pipeline does NOT support single-end data. From the command-line, each input file should separated by a space. Multiple input FastQ files per sample can be provided. Globbing is supported! This makes selecting FastQ files easy. Input FastQ files should always be gzipp-ed.</p> <p><strong><em>Example:</em></strong> <code>--input .tests/*.R?.fastq.gz</code></p> <p>Cell Ranger Input: Cell Ranger output folders can be provided. It is expected that the outs folder is contained within the Cell Ranger output folders, and keep the normal output folder structure. Globbing is supported!</p> <p><strong><em>Example:</em></strong> <code>--input .tests/*/</code></p> </blockquote> <hr> <p><code>--output OUTPUT</code></p> <blockquote> <p><strong>Path to an output directory.</strong> <br> <em>type: path</em></p> <p>This location is where the pipeline will create all of its output files, also known as the pipeline's working directory. If the provided output directory does not exist, it will be created automatically.</p> <p><strong><em>Example:</em></strong> <code>--output /data/$USER/cell-seek_out</code></p> </blockquote> <hr> <p><code>--pipeline multiome</code></p> <blockquote> <p><strong>The pipeline to run.</strong> <br> <em>type: string</em></p> <p>This option selects the version of the pipeline to run. The documentation provided is based on choosing the option for multiome.</p> <p><strong><em>Example:</em></strong> <code>--pipeline multiome</code></p> </blockquote> <hr> <p><code>--genome {hg38, mm10, hg2024, mm2024, custom.json}</code></p> <blockquote> <p><strong>Reference genome.</strong> <br> <em>type: string</em></p> <p>This option defines the reference genome of the samples. cell-seek does comes bundled with prebuilt reference files for human and mouse samples, e.g. hg38 or mm10.</p> <p>A custom reference genome can also be provided via a json file. Additional information for creating this json file can be found in <a href=../genome/ ><code>cell-seek <b>genome</b></code></a>.</p> <p>For prebuilt references please select one of the following options: hg38, mm10</p> <p><strong><em>Example:</em></strong> <code>--genome hg38</code></p> </blockquote> <h4 id=262-conditionally-required-arguments>2.6.2 Conditionally Required Arguments<a class=headerlink href=#262-conditionally-required-arguments title="Permanent link">&para;</a></h4> <p>The following arguments are only required when FastQ files are used as input. They are not required when Cell Ranger output file is used as input.</p> <p><code>--libraries LIBRARIES</code></p> <blockquote> <p><strong>Libraries file.</strong> <br> <em>type: file</em></p> <p>A CSV file containing information about each library. It contains each sample's name, flowcell, demultiplexed name, and library type. More information about the libraries file and its requirements can be found on the <a href=https://support.10xgenomics.com/single-cell-multiome-atac-gex/software/pipelines/latest/using/count#library-csv>10x Genomics website</a>.</p> <p><em>Here is an example libraries.csv file:</em> <div class=highlight><pre><span></span><code>Name,Flowcell,Sample,Type
IL15_LNs,H7CNNBGXG,IL15_LNs,Gene Expression
IL15_LNs,H7CT7BGXG,IL15_LNs_BC,Antibody Capture
</code></pre></div></p> <p><em>Where:</em></p> <ul> <li><em>Name:</em> name of the sample passed to Cell Ranger. </li> <li><em>Flowcell:</em> The flowcell ID that contains the FASTQ files for this set of data. </li> <li><em>Sample:</em> Name that was used when demultiplexing, this should match the FASTQ files. </li> <li><em>Type:</em> library type for each sample. List of supported options: <ul> <li>Gene Expression</li> <li>Chromatin Accessibility</li> </ul> </li> </ul> <p><strong><em>Example:</em></strong> <code>--libraries libraries.csv</code></p> </blockquote> <h4 id=263-analysis-options>2.6.3 Analysis Options<a class=headerlink href=#263-analysis-options title="Permanent link">&para;</a></h4> <p>The multiome pipeline currently does not have any applicable analysis flags.</p> <h3 id=27-orchestration-options>2.7 Orchestration Options<a class=headerlink href=#27-orchestration-options title="Permanent link">&para;</a></h3> <p>Each of the following arguments are optional, and do not need to be provided.</p> <p><code>--dry-run</code> </p> <blockquote> <p><strong>Dry run the pipeline.</strong><br> <em>type: boolean flag</em></p> <p>Displays what steps in the pipeline remain or will be run. Does not execute anything!</p> <p><strong><em>Example:</em></strong> <code>--dry-run</code></p> </blockquote> <hr> <p><code>--silent</code> </p> <blockquote> <p><strong>Silence standard output.</strong><br> <em>type: boolean flag</em></p> <p>Reduces the amount of information directed to standard output when submitting master job to the job scheduler. Only the job id of the master job is returned.</p> <p><strong><em>Example:</em></strong> <code>--silent</code></p> </blockquote> <hr> <p><code>--mode {slurm,local}</code> </p> <blockquote> <p><strong>Execution Method.</strong><br> <em>type: string</em><br> <em>default: slurm</em></p> <p>Execution Method. Defines the mode or method of execution. Vaild mode options include: slurm or local.</p> <p><strong><em>slurm</em></strong> <br> The slurm execution method will submit jobs to the <a href=https://slurm.schedmd.com/ >SLURM workload manager</a>. It is recommended running cell-seek in this mode as execution will be significantly faster in a distributed environment. This is the default mode of execution.</p> <p><strong><em>local</em></strong><br> Local executions will run serially on compute instance. This is useful for testing, debugging, or when a users does not have access to a high performance computing environment. If this option is not provided, it will default to a local execution mode.</p> <p><strong><em>Example:</em></strong> <code>--mode slurm</code></p> </blockquote> <hr> <p><code>--job-name JOB_NAME</code> </p> <blockquote> <p><strong>Set the name of the pipeline's master job.</strong><br> <em>type: string</em> <em>default: pl:cell-seek</em></p> <p>When submitting the pipeline to a job scheduler, like SLURM, this option always you to set the name of the pipeline's master job. By default, the name of the pipeline's master job is set to "pl:cell-seek".</p> <p><strong><em>Example:</em></strong> <code>--job-name pl_id-42</code></p> </blockquote> <hr> <p><code>--singularity-cache SINGULARITY_CACHE</code> </p> <blockquote> <p><strong>Overrides the $SINGULARITY_CACHEDIR environment variable.</strong><br> <em>type: path</em><br> <em>default: <code>--output OUTPUT/.singularity</code></em></p> <p>Singularity will cache image layers pulled from remote registries. This ultimately speeds up the process of pull an image from DockerHub if an image layer already exists in the singularity cache directory. By default, the cache is set to the value provided to the <code>--output</code> argument. Please note that this cache cannot be shared across users. Singularity strictly enforces you own the cache directory and will return a non-zero exit code if you do not own the cache directory! See the <code>--sif-cache</code> option to create a shareable resource.</p> <p><strong><em>Example:</em></strong> <code>--singularity-cache /data/$USER/.singularity</code></p> </blockquote> <hr> <p><code>--sif-cache SIF_CACHE</code></p> <blockquote> <p><strong>Path where a local cache of SIFs are stored.</strong><br> <em>type: path</em> </p> <p>Uses a local cache of SIFs on the filesystem. This SIF cache can be shared across users if permissions are set correctly. If a SIF does not exist in the SIF cache, the image will be pulled from Dockerhub and a warning message will be displayed. The <code>cell-seek cache</code> subcommand can be used to create a local SIF cache. Please see <code>cell-seek cache</code> for more information. This command is extremely useful for avoiding DockerHub pull rate limits. It also remove any potential errors that could occur due to network issues or DockerHub being temporarily unavailable. We recommend running cell-seek with this option when ever possible.</p> <p><strong><em>Example:</em></strong> <code>--singularity-cache /data/$USER/SIFs</code></p> </blockquote> <hr> <p><code>--threads THREADS</code> </p> <blockquote> <p><strong>Max number of threads for each process.</strong><br> <em>type: int</em><br> <em>default: 2</em></p> <p>Max number of threads for each process. This option is more applicable when running the pipeline with <code>--mode local</code>. It is recommended setting this vaule to the maximum number of CPUs available on the host machine.</p> <p><strong><em>Example:</em></strong> <code>--threads 12</code></p> </blockquote> <hr> <p><code>--tmp-dir TMP_DIR</code> </p> <blockquote> <p><strong>Max number of threads for each process.</strong><br> <em>type: path</em><br> <em>default: <code>/lscratch/$SLURM_JOBID</code></em></p> <p>Path on the file system for writing temporary output files. By default, the temporary directory is set to '/lscratch/$SLURM_JOBID' for backwards compatibility with the NIH's Biowulf cluster; however, if you are running the pipeline on another cluster, this option will need to be specified. Ideally, this path should point to a dedicated location on the filesystem for writing tmp files. On many systems, this location is set to somewhere in /scratch. If you need to inject a variable into this string that should NOT be expanded, please quote this options value in single quotes.</p> <p><strong><em>Example:</em></strong> <code>--tmp-dir /scratch/$USER/</code></p> </blockquote> <h3 id=28-miscellaneous-options>2.8 Miscellaneous Options<a class=headerlink href=#28-miscellaneous-options title="Permanent link">&para;</a></h3> <p>Each of the following arguments are optional, and do not need to be provided.</p> <p><code>-h, --help</code> </p> <blockquote> <p><strong>Display Help.</strong><br> <em>type: boolean flag</em></p> <p>Shows command's synopsis, help message, and an example command</p> <p><strong><em>Example:</em></strong> <code>--help</code></p> </blockquote> <h2 id=3-example>3. Example<a class=headerlink href=#3-example title="Permanent link">&para;</a></h2> <h3 id=31-gex-example>3.1 GEX Example<a class=headerlink href=#31-gex-example title="Permanent link">&para;</a></h3> <div class=highlight><pre><span></span><code><span class=c1># Step 1.) Grab an interactive node,</span>
<span class=c1># do not run on head node!</span>
srun<span class=w> </span>-N<span class=w> </span><span class=m>1</span><span class=w> </span>-n<span class=w> </span><span class=m>1</span><span class=w> </span>--time<span class=o>=</span><span class=m>1</span>:00:00<span class=w> </span>--mem<span class=o>=</span>8gb<span class=w>  </span>--cpus-per-task<span class=o>=</span><span class=m>2</span><span class=w> </span>--pty<span class=w> </span>bash
module<span class=w> </span>purge
module<span class=w> </span>load<span class=w> </span>singularity<span class=w> </span>snakemake

<span class=c1># Step 2A.) Dry-run the pipeline</span>
./cell-seek<span class=w> </span>run<span class=w> </span>--input<span class=w> </span>.tests/*.R?.fastq.gz<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--output<span class=w> </span>/data/<span class=nv>$USER</span>/output<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--pipeline<span class=w> </span>gex<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--genome<span class=w> </span>hg38<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--cellranger<span class=w> </span><span class=m>8</span>.0.0<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--mode<span class=w> </span>slurm<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--dry-run

<span class=c1># Step 2B.) Run the cell-seek pipeline</span>
<span class=c1># The slurm mode will submit jobs to</span>
<span class=c1># the cluster. It is recommended running</span>
<span class=c1># the pipeline in this mode.</span>
./cell-seek<span class=w> </span>run<span class=w> </span>--input<span class=w> </span>.tests/*.R?.fastq.gz<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--output<span class=w> </span>/data/<span class=nv>$USER</span>/output<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--pipeline<span class=w> </span>gex<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--genome<span class=w> </span>hg38<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--cellranger<span class=w> </span><span class=m>8</span>.0.0<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--mode<span class=w> </span>slurm
</code></pre></div> <h3 id=32-vdj-example>3.2 VDJ Example<a class=headerlink href=#32-vdj-example title="Permanent link">&para;</a></h3> <div class=highlight><pre><span></span><code><span class=c1># Step 1.) Grab an interactive node,</span>
<span class=c1># do not run on head node!</span>
srun<span class=w> </span>-N<span class=w> </span><span class=m>1</span><span class=w> </span>-n<span class=w> </span><span class=m>1</span><span class=w> </span>--time<span class=o>=</span><span class=m>1</span>:00:00<span class=w> </span>--mem<span class=o>=</span>8gb<span class=w>  </span>--cpus-per-task<span class=o>=</span><span class=m>2</span><span class=w> </span>--pty<span class=w> </span>bash
module<span class=w> </span>purge
module<span class=w> </span>load<span class=w> </span>singularity<span class=w> </span>snakemake

<span class=c1># Step 2A.) Dry-run the pipeline</span>
./cell-seek<span class=w> </span>run<span class=w> </span>--input<span class=w> </span>.tests/*.R?.fastq.gz<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--output<span class=w> </span>/data/<span class=nv>$USER</span>/output<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--pipeline<span class=w> </span>vdj<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--genome<span class=w> </span>hg38<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--cellranger<span class=w> </span><span class=m>8</span>.0.0<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--mode<span class=w> </span>slurm<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--dry-run

<span class=c1># Step 2B.) Run the cell-seek pipeline</span>
<span class=c1># The slurm mode will submit jobs to</span>
<span class=c1># the cluster. It is recommended running</span>
<span class=c1># the pipeline in this mode.</span>
./cell-seek<span class=w> </span>run<span class=w> </span>--input<span class=w> </span>.tests/*.R?.fastq.gz<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--output<span class=w> </span>/data/<span class=nv>$USER</span>/output<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--pipeline<span class=w> </span>vdj<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--genome<span class=w> </span>hg38<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--cellranger<span class=w> </span><span class=m>8</span>.0.0<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--mode<span class=w> </span>slurm
</code></pre></div> <h3 id=33-cite-example>3.3 CITE Example<a class=headerlink href=#33-cite-example title="Permanent link">&para;</a></h3> <div class=highlight><pre><span></span><code><span class=c1># Step 1.) Grab an interactive node,</span>
<span class=c1># do not run on head node!</span>
srun<span class=w> </span>-N<span class=w> </span><span class=m>1</span><span class=w> </span>-n<span class=w> </span><span class=m>1</span><span class=w> </span>--time<span class=o>=</span><span class=m>1</span>:00:00<span class=w> </span>--mem<span class=o>=</span>8gb<span class=w>  </span>--cpus-per-task<span class=o>=</span><span class=m>2</span><span class=w> </span>--pty<span class=w> </span>bash
module<span class=w> </span>purge
module<span class=w> </span>load<span class=w> </span>singularity<span class=w> </span>snakemake

<span class=c1># Step 2A.) Dry-run the pipeline</span>
./cell-seek<span class=w> </span>run<span class=w> </span>--input<span class=w> </span>.tests/*.R?.fastq.gz<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--output<span class=w> </span>/data/<span class=nv>$USER</span>/output<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--pipeline<span class=w> </span>cite<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--genome<span class=w> </span>hg38<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--cellranger<span class=w> </span><span class=m>8</span>.0.0<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--libraries<span class=w> </span>libraries.csv<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--features<span class=w> </span>features.csv<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--mode<span class=w> </span>slurm<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--dry-run

<span class=c1># Step 2B.) Run the cell-seek pipeline</span>
<span class=c1># The slurm mode will submit jobs to</span>
<span class=c1># the cluster. It is recommended running</span>
<span class=c1># the pipeline in this mode.</span>
./cell-seek<span class=w> </span>run<span class=w> </span>--input<span class=w> </span>.tests/*.R?.fastq.gz<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--output<span class=w> </span>/data/<span class=nv>$USER</span>/output<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--pipeline<span class=w> </span>cite<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--genome<span class=w> </span>hg38<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--cellranger<span class=w> </span><span class=m>8</span>.0.0<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--libraries<span class=w> </span>libraries.csv<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--features<span class=w> </span>features.csv<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--mode<span class=w> </span>slurm
</code></pre></div> <h3 id=34-multi-example>3.4 MULTI Example<a class=headerlink href=#34-multi-example title="Permanent link">&para;</a></h3> <h4 id=341-gex-and-vdj>3.4.1 GEX and VDJ<a class=headerlink href=#341-gex-and-vdj title="Permanent link">&para;</a></h4> <p>The following is an example of running <code>multi</code> when combining GEX and VDJ runs.</p> <div class=highlight><pre><span></span><code><span class=c1># Step 1.) Grab an interactive node,</span>
<span class=c1># do not run on head node!</span>
srun<span class=w> </span>-N<span class=w> </span><span class=m>1</span><span class=w> </span>-n<span class=w> </span><span class=m>1</span><span class=w> </span>--time<span class=o>=</span><span class=m>1</span>:00:00<span class=w> </span>--mem<span class=o>=</span>8gb<span class=w>  </span>--cpus-per-task<span class=o>=</span><span class=m>2</span><span class=w> </span>--pty<span class=w> </span>bash
module<span class=w> </span>purge
module<span class=w> </span>load<span class=w> </span>singularity<span class=w> </span>snakemake

<span class=c1># Step 2A.) Dry-run the pipeline</span>
./cell-seek<span class=w> </span>run<span class=w> </span>--input<span class=w> </span>.tests/*.R?.fastq.gz<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--output<span class=w> </span>/data/<span class=nv>$USER</span>/output<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--pipeline<span class=w> </span>multi<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--genome<span class=w> </span>hg38<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--cellranger<span class=w> </span><span class=m>8</span>.0.0<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--libraries<span class=w> </span>libraries.csv<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--mode<span class=w> </span>slurm<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--dry-run

<span class=c1># Step 2B.) Run the cell-seek pipeline</span>
<span class=c1># The slurm mode will submit jobs to</span>
<span class=c1># the cluster. It is recommended running</span>
<span class=c1># the pipeline in this mode.</span>
./cell-seek<span class=w> </span>run<span class=w> </span>--input<span class=w> </span>.tests/*.R?.fastq.gz<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--output<span class=w> </span>/data/<span class=nv>$USER</span>/output<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--pipeline<span class=w> </span>multi<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--genome<span class=w> </span>hg38<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--cellranger<span class=w> </span><span class=m>8</span>.0.0<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--libraries<span class=w> </span>libraries.csv<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--mode<span class=w> </span>slurm
</code></pre></div> <h4 id=342-including-hto>3.4.2 Including HTO<a class=headerlink href=#342-including-hto title="Permanent link">&para;</a></h4> <p>The following is an example of running <code>multi</code> while providing hashtag information.</p> <div class=highlight><pre><span></span><code><span class=c1># Step 1.) Grab an interactive node,</span>
<span class=c1># do not run on head node!</span>
srun<span class=w> </span>-N<span class=w> </span><span class=m>1</span><span class=w> </span>-n<span class=w> </span><span class=m>1</span><span class=w> </span>--time<span class=o>=</span><span class=m>1</span>:00:00<span class=w> </span>--mem<span class=o>=</span>8gb<span class=w>  </span>--cpus-per-task<span class=o>=</span><span class=m>2</span><span class=w> </span>--pty<span class=w> </span>bash
module<span class=w> </span>purge
module<span class=w> </span>load<span class=w> </span>singularity<span class=w> </span>snakemake

<span class=c1># Step 2A.) Dry-run the pipeline</span>
./cell-seek<span class=w> </span>run<span class=w> </span>--input<span class=w> </span>.tests/*.R?.fastq.gz<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--output<span class=w> </span>/data/<span class=nv>$USER</span>/output<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--pipeline<span class=w> </span>multi<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--genome<span class=w> </span>hg38<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--cellranger<span class=w> </span><span class=m>8</span>.0.0<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--libraries<span class=w> </span>libraries.csv<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--cmo-reference<span class=w> </span>cmo_reference.csv<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--mode<span class=w> </span>slurm<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--dry-run

<span class=c1># Step 2B.) Run the cell-seek pipeline</span>
<span class=c1># The slurm mode will submit jobs to</span>
<span class=c1># the cluster. It is recommended running</span>
<span class=c1># the pipeline in this mode.</span>
./cell-seek<span class=w> </span>run<span class=w> </span>--input<span class=w> </span>.tests/*.R?.fastq.gz<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--output<span class=w> </span>/data/<span class=nv>$USER</span>/output<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--pipeline<span class=w> </span>multi<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--genome<span class=w> </span>hg38<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--cellranger<span class=w> </span><span class=m>8</span>.0.0<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--libraries<span class=w> </span>libraries.csv<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--cmo-reference<span class=w> </span>cmo_reference.csv<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--mode<span class=w> </span>slurm
</code></pre></div> <h3 id=35-atac-example>3.5 ATAC Example<a class=headerlink href=#35-atac-example title="Permanent link">&para;</a></h3> <div class=highlight><pre><span></span><code><span class=c1># Step 1.) Grab an interactive node,</span>
<span class=c1># do not run on head node!</span>
srun<span class=w> </span>-N<span class=w> </span><span class=m>1</span><span class=w> </span>-n<span class=w> </span><span class=m>1</span><span class=w> </span>--time<span class=o>=</span><span class=m>1</span>:00:00<span class=w> </span>--mem<span class=o>=</span>8gb<span class=w>  </span>--cpus-per-task<span class=o>=</span><span class=m>2</span><span class=w> </span>--pty<span class=w> </span>bash
module<span class=w> </span>purge
module<span class=w> </span>load<span class=w> </span>singularity<span class=w> </span>snakemake

<span class=c1># Step 2A.) Dry-run the pipeline</span>
./cell-seek<span class=w> </span>run<span class=w> </span>--input<span class=w> </span>.tests/*.R?.fastq.gz<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--output<span class=w> </span>/data/<span class=nv>$USER</span>/output<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--pipeline<span class=w> </span>atac<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--genome<span class=w> </span>hg38<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--mode<span class=w> </span>slurm<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--dry-run

<span class=c1># Step 2B.) Run the cell-seek pipeline</span>
<span class=c1># The slurm mode will submit jobs to</span>
<span class=c1># the cluster. It is recommended running</span>
<span class=c1># the pipeline in this mode.</span>
./cell-seek<span class=w> </span>run<span class=w> </span>--input<span class=w> </span>.tests/*.R?.fastq.gz<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--output<span class=w> </span>/data/<span class=nv>$USER</span>/output<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--pipeline<span class=w> </span>atac<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--genome<span class=w> </span>hg38<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--mode<span class=w> </span>slurm
</code></pre></div> <h3 id=36-multiome-example>3.6 Multiome Example<a class=headerlink href=#36-multiome-example title="Permanent link">&para;</a></h3> <div class=highlight><pre><span></span><code><span class=c1># Step 1.) Grab an interactive node,</span>
<span class=c1># do not run on head node!</span>
srun<span class=w> </span>-N<span class=w> </span><span class=m>1</span><span class=w> </span>-n<span class=w> </span><span class=m>1</span><span class=w> </span>--time<span class=o>=</span><span class=m>1</span>:00:00<span class=w> </span>--mem<span class=o>=</span>8gb<span class=w>  </span>--cpus-per-task<span class=o>=</span><span class=m>2</span><span class=w> </span>--pty<span class=w> </span>bash
module<span class=w> </span>purge
module<span class=w> </span>load<span class=w> </span>singularity<span class=w> </span>snakemake

<span class=c1># Step 2A.) Dry-run the pipeline</span>
./cell-seek<span class=w> </span>run<span class=w> </span>--input<span class=w> </span>.tests/*.R?.fastq.gz<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--output<span class=w> </span>/data/<span class=nv>$USER</span>/output<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--pipeline<span class=w> </span>multiome<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--genome<span class=w> </span>hg38<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--libraries<span class=w> </span>libraries.csv<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--mode<span class=w> </span>slurm<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--dry-run

<span class=c1># Step 2B.) Run the cell-seek pipeline</span>
<span class=c1># The slurm mode will submit jobs to</span>
<span class=c1># the cluster. It is recommended running</span>
<span class=c1># the pipeline in this mode.</span>
./cell-seek<span class=w> </span>run<span class=w> </span>--input<span class=w> </span>.tests/*.R?.fastq.gz<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--output<span class=w> </span>/data/<span class=nv>$USER</span>/output<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--pipeline<span class=w> </span>multiome<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--genome<span class=w> </span>hg38<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--libraries<span class=w> </span>libraries.csv<span class=w> </span><span class=se>\</span>
<span class=w>                  </span>--mode<span class=w> </span>slurm
</code></pre></div> <aside class=md-source-file> <span class=md-source-file__fact> <span class=md-icon title="Last update"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M21 13.1c-.1 0-.3.1-.4.2l-1 1 2.1 2.1 1-1c.2-.2.2-.6 0-.8l-1.3-1.3c-.1-.1-.2-.2-.4-.2m-1.9 1.8-6.1 6V23h2.1l6.1-6.1zM12.5 7v5.2l4 2.4-1 1L11 13V7zM11 21.9c-5.1-.5-9-4.8-9-9.9C2 6.5 6.5 2 12 2c5.3 0 9.6 4.1 10 9.3-.3-.1-.6-.2-1-.2s-.7.1-1 .2C19.6 7.2 16.2 4 12 4c-4.4 0-8 3.6-8 8 0 4.1 3.1 7.5 7.1 7.9l-.1.2z"/></svg> </span> 2026-01-08 </span> </aside> </article> </div> <script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script> </div> <button type=button class="md-top md-icon" data-md-component=top hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"/></svg> Back to top </button> </main> <footer class=md-footer> <div class="md-footer-meta md-typeset"> <div class="md-footer-meta__inner md-grid"> <div class=md-copyright> <div class=md-copyright__highlight> Copyright &copy; 2026 OpenOmics </div> Made with <a href=https://squidfunk.github.io/mkdocs-material/ target=_blank rel=noopener> Material for MkDocs </a> </div> <div class=md-social> <a href=https://ncbr.ncifcrf.gov/ target=_blank rel=noopener title=ncbr.ncifcrf.gov class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 640 512"><!-- Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M320 16a104 104 0 1 1 0 208 104 104 0 1 1 0-208M96 88a72 72 0 1 1 0 144 72 72 0 1 1 0-144M0 416c0-70.7 57.3-128 128-128 12.8 0 25.2 1.9 36.9 5.4C132 330.2 112 378.8 112 432v16c0 11.4 2.4 22.2 6.7 32H32c-17.7 0-32-14.3-32-32zm521.3 64c4.3-9.8 6.7-20.6 6.7-32v-16c0-53.2-20-101.8-52.9-138.6 11.7-3.5 24.1-5.4 36.9-5.4 70.7 0 128 57.3 128 128v32c0 17.7-14.3 32-32 32zM472 160a72 72 0 1 1 144 0 72 72 0 1 1-144 0M160 432c0-88.4 71.6-160 160-160s160 71.6 160 160v16c0 17.7-14.3 32-32 32H192c-17.7 0-32-14.3-32-32z"/></svg> </a> <a href=https://github.com/OpenOmics target=_blank rel=noopener title=github.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 512 512"><!-- Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M173.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M252.8 8C114.1 8 8 113.3 8 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C436.2 457.8 504 362.9 504 252 504 113.3 391.5 8 252.8 8M105.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg> </a> <a href=https://hub.docker.com/u/skchronicles target=_blank rel=noopener title=hub.docker.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 640 512"><!-- Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M349.9 236.3h-66.1v-59.4h66.1zm0-204.3h-66.1v60.7h66.1zm78.2 144.8H362v59.4h66.1zm-156.3-72.1h-66.1v60.1h66.1zm78.1 0h-66.1v60.1h66.1zm276.8 100c-14.4-9.7-47.6-13.2-73.1-8.4-3.3-24-16.7-44.9-41.1-63.7l-14-9.3-9.3 14c-18.4 27.8-23.4 73.6-3.7 103.8-8.7 4.7-25.8 11.1-48.4 10.7H2.4c-8.7 50.8 5.8 116.8 44 162.1 37.1 43.9 92.7 66.2 165.4 66.2 157.4 0 273.9-72.5 328.4-204.2 21.4.4 67.6.1 91.3-45.2 1.5-2.5 6.6-13.2 8.5-17.1zm-511.1-27.9h-66v59.4h66.1v-59.4zm78.1 0h-66.1v59.4h66.1zm78.1 0h-66.1v59.4h66.1zm-78.1-72.1h-66.1v60.1h66.1z"/></svg> </a> </div> </div> </div> </footer> </div> <div class=md-dialog data-md-component=dialog> <div class="md-dialog__inner md-typeset"></div> </div> <script id=__config type=application/json>{"annotate": null, "base": "../..", "features": ["navigation.tabs", "navigation.top", "toc.integrate"], "search": "../../assets/javascripts/workers/search.2c215733.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": {"provider": "mike"}}</script> <script src=../../assets/javascripts/bundle.79ae519e.min.js></script> </body> </html>